// light_tx_history/light_tx_history.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { sdb, get } = statedb(fallback_module)
const light_tx_row = require('light_tx_row')

module.exports = light_tx_history

async function light_tx_history(opts = {}, { onClose } = {}) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb
  const on = { style: inject, data: ondata }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  shadow.innerHTML = `
    <div class="transaction-history">
      <div class="transaction-history-header">
        <span>Transaction History</span>
        <button class="history-close">✕</button>
      </div>
      <div class="transaction-history-body"></div>
    </div>
    <style></style>
  `

  const style = shadow.querySelector('style')
  const container = shadow.querySelector('.transaction-history')
  const body = shadow.querySelector('.transaction-history-body')
  const closeBtn = shadow.querySelector('.history-close')

  const subs = await sdb.watch(onbatch)

  // hide overlay on close
  closeBtn.addEventListener('click', () => {
    if (onClose) onClose()
  })

  // build grouped list
  const grouped = {}
  subs.forEach(sub => {
    const date = (sub.date || 'Unknown').trim()
    if (!grouped[date]) grouped[date] = []
    grouped[date].push(sub)
  })

  for (const date in grouped) {
    const dateEl = document.createElement('div')
    dateEl.className = 'transaction-date'
    dateEl.textContent = date
    body.appendChild(dateEl)

    for (const sub of grouped[date]) {
      const row = await light_tx_row(sub)
      body.appendChild(row)
    }
  }

  return el

  function fail(data, type) {
    throw new Error('invalid message', { cause: { data, type } })
  }

  async function onbatch(batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(
        paths.map(p => drive.get(p).then(f => f.raw))
      )
      const fn = on[type] || fail
      await fn(data)
    }
  }

  function inject(data) {
    style.textContent = data[0]
  }

  async function ondata() {}
}

function fallback_module () {
  return {
    api,
    _: {
      'light_tx_row':{
        $: ''
      }
    } 
  }
  function api(opts){
    if (!opts) opts = {}
    if (!opts.value) opts.value = {}

    const light_tx_row = {
      mapping: {
        style: 'style',
        data: 'data'
      }
    }
    opts.value.forEach((transaction, index) => {
      light_tx_row[index] = transaction
    })
    return {
      drive: {
        'style/': {
          'style.css':{
            '$ref': 'light_tx_history.css'
          }
        },
        'data/': {
          'opts.json':{
            raw: opts || {}
          }
        }
      },
      _:{
        light_tx_row
      }
    }
  }
}
