// add_contact_popup/add_contact_popup.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { sdb, get } = statedb(fallback_module)

const gen_invite_code = require('gen_invite_code')
const add_new_contact = require('add_new_contact')

module.exports = add_contact_popup

async function add_contact_popup (opts = {}, { onClose } = {}) {
  const { sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = {
    style: inject,
    data: ondata,
    icons: iconject,
  }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  let icons = []
  let show_component = false 

  shadow.innerHTML = `
    <div class="more-menu-container"></div>
    <div class="gen-invite hidden"></div>
    <div class="add-new hidden"></div>
    <style></style>
  `
  el.classList.add('position')

  const style = shadow.querySelector('style')
  const row = shadow.querySelector('.more-menu-container')
  const gen_invite_el = shadow.querySelector('.gen-invite')
  const add_new_el = shadow.querySelector('.add-new')

  const subs = await sdb.watch(onbatch)

  let invite_mounted = false
  let add_new_mounted = false

  return el

  async function onbatch(batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(
        paths.map(path => {
          return drive.get(path).then(file => file.raw)
        })
      )
      const func = on[type] || fail
      await func(data, type)
    }
  }

  function inject(data) {
    style.textContent = data[0]
  }

  async function ondata() {

    if (show_component) return

    row.innerHTML = `
      <div class="container-title">
        <div class="title">Add Contact</div>
        <div class="close-icon">${icons[0]}</div>
      </div>

      <div class="option-container" id="invite-option">
        <div class="dot-icon">${icons[1]}</div>
        <div class="option-label">Generate Invite Code</div>
      </div>
      <div class="option-container" id="add-new-option">
        <div class="dot-icon">${icons[2]}</div>
        <div class="option-label">Add new contact</div>
      </div>
    `

    const close_btn = row.querySelector('.close-icon')
    const invite_btn = row.querySelector('#invite-option')
    const add_new_btn = row.querySelector('#add-new-option')

    // close button functionality
    close_btn.addEventListener('click', () => {
      if (onClose) onClose()
    })

    // invite button functionality
    invite_btn.addEventListener('click', async () => {
      el.classList.remove('position')
      gen_invite_el.classList.remove('hidden')

      if (!invite_mounted) {
        const gen_invite_comp = await gen_invite_code(subs[0],{ 
          onClose: () => {
            el.classList.add('position')
            gen_invite_el.classList.add('hidden')

          }})
        gen_invite_el.append(gen_invite_comp)
        invite_mounted = true
      }
    })

    // add new button functionality
    add_new_btn.addEventListener('click', async () => {
      el.classList.remove('position')
      add_new_el.classList.remove('hidden')

      if (!add_new_mounted){
        const add_new_comp = await add_new_contact(subs[1],{
          onClose: () => {
            el.classList.add('position')
            add_new_el.classList.add('hidden')
          }})
        add_new_el.append(add_new_comp)
        add_new_mounted = true
      }
    })
  }

  function fail(data, type) {
    console.error('[FAIL]', data, type)
    throw new Error('invalid message', { cause: { data, type } })
  }

  function iconject(data) {
    icons = data
  }
}


function fallback_module () {
  return {
    api: fallback_instance,
    _: {
      'gen_invite_code': { $: '' },
      'add_new_contact': { $: '' },
    }
  }

  function fallback_instance (opts) {
    if (!opts) opts = {}
    if (!opts.value) opts.value = {}

    const gen_invite_code = {
      mapping: { style: 'style', data: 'data', icons: 'icons' },
      0: {}
    }

    const add_new_contact = {
      mapping: { style: 'style', data: 'data', icons: 'icons' },
      1: {}
    }

    return {
      drive: {
        'icons/': {
          'x.svg': { '$ref': 'x.svg' },
          'link.svg': { '$ref': 'link.svg' },
          'user_add.svg': { '$ref': 'user_add.svg' },
        },
        'style/': {
          'style.css': { '$ref': 'add_contact_popup.css' },
        },
        'data/': {
          'opts.json': {
            raw: opts || {}
          },
        }
      },
      _: { gen_invite_code, add_new_contact }
    }
  }
}
