// add_contact_popup/add_contact_popup.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { sdb, get } = statedb(fallback_module)

const gen_invite_code = require('gen_invite_code')
const add_new_contact = require('add_new_contact')

module.exports = add_contact_popup

async function add_contact_popup (opts = {}) {

  const { sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = {
    style: inject,
    data: ondata,
    icons: iconject,
  }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  let icons = []
  let invitePopup = null
  let addContactPopup = null

  shadow.innerHTML = `
    <div class="more-menu-container"></div>
    <div class="popup"></div>
    <style></style>
  `

  const style = shadow.querySelector('style')
  const row = shadow.querySelector('.more-menu-container')
  const popup = shadow.querySelector('.popup')

  const subs = await sdb.watch(onbatch)

  const gen_invite_code_comp = await gen_invite_code(subs[0])
  const add_new_contact_comp = await add_new_contact(subs[1])

  return el

  async function onbatch(batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(
        paths.map(path => {
          return drive.get(path).then(file => file.raw)
        })
      )
      const func = on[type] || fail
      await func(data, type)
    }
  }

  function inject(data) {
    style.textContent = data[0]
  }

  async function ondata() {
    row.innerHTML = `
      <div class="container-title">
        <div class="title">Add Contact</div>
        <div class="close-icon">${icons[0]}</div>
      </div>

      <div class="option-container" id="invite-option">
        <div class="dot-icon">${icons[1]}</div>
        <div class="option-label">Generate Invite Code</div>
      </div>
      <div class="option-container" id="add-contact-option">
        <div class="dot-icon">${icons[2]}</div>
        <div class="option-label">Add new contact</div>
      </div>
    `

    const closeBtn = row.querySelector('.close-icon')
    const inviteBtn = row.querySelector('#invite-option')
    const addContactBtn = row.querySelector('#add-contact-option')

    if (closeBtn) {
      closeBtn.onclick = () => {
        const dropdown = el.parentNode
        if (dropdown) dropdown.classList.add('hidden')
      }
    }

    // ✅ OPTION 1 — TOGGLE GEN INVITE
    inviteBtn.onclick = async () => {
      console.log('[INVITE CLICKED]')

      // ✅ If already open → CLOSE IT
      if (invitePopup) {
        console.log('[INVITE] Closing')
        invitePopup.remove()
        invitePopup = null
        return
      }

      // ✅ If other popup open → CLOSE IT
      if (addContactPopup) {
        console.log('[INVITE] Closing add contact first')
        addContactPopup.remove()
        addContactPopup = null
      }

      console.log('[INVITE] Opening')
      invitePopup = gen_invite_code_comp
      popup.appendChild(invitePopup)
    }

    // ✅ OPTION 2 — TOGGLE ADD CONTACT
    addContactBtn.onclick = async () => {
      console.log('[ADD CONTACT CLICKED]')

      // ✅ If already open → CLOSE IT
      if (addContactPopup) {
        console.log('[ADD CONTACT] Closing')
        addContactPopup.remove()
        addContactPopup = null
        return
      }

      // ✅ If other popup open → CLOSE IT
      if (invitePopup) {
        console.log('[ADD CONTACT] Closing invite first')
        invitePopup.remove()
        invitePopup = null
      }

      console.log('[ADD CONTACT] Opening')
      addContactPopup = add_new_contact_comp
      popup.appendChild(addContactPopup)
    }
  }

  function fail(data, type) {
    console.error('[FAIL]', data, type)
    throw new Error('invalid message', { cause: { data, type } })
  }

  function iconject(data) {
    icons = data
  }
}


function fallback_module () {
  return {
    api: fallback_instance,
    _: {
      'gen_invite_code': { $: '' },
      'add_new_contact': { $: '' },
    }
  }

  function fallback_instance (opts) {

    const gen_invite_code = {
      mapping: { style: 'style', data: 'data', icons: 'icons' },
      0: {}
    }

    const add_new_contact = {
      mapping: { style: 'style', data: 'data', icons: 'icons' },
      1: {}
    }

    return {
      drive: {
        'icons/': {
          'x.svg': { '$ref': 'x.svg' },
          'link.svg': { '$ref': 'link.svg' },
          'user_add.svg': { '$ref': 'user_add.svg' },
        },
        'style/': {
          'style.css': { '$ref': 'add_contact_popup.css' },
        },
        'data/': {
          'opts.json': {
            raw: opts
          },
        }
      },
      _: { gen_invite_code, add_new_contact }
    }
  }
}
