// home_contents/home_contents.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { sdb, get } = statedb(fallback_module)

const home_page_header = require('home_page_header')
const action_buttons = require('action_buttons')
const transaction_list = require('transaction_list')
const total_wealth = require('total_wealth')
const lightning_buttons = require('lightning_buttons')
const wallet_button = require('wallet_button')
const light_page_header = require('light_page_header')
const light_tx_list = require('light_tx_list')

module.exports = home_contents

async function home_contents(opts = {}) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = { style: inject, data: ondata }

  const el = document.createElement('div')
  el.id = "home_contents"
  const shadow = el.attachShadow({ mode: 'closed' })

  shadow.innerHTML = `
    <div class="home-contents-container"></div>
    <style></style>
  `

  const style = shadow.querySelector('style')
  const container = shadow.querySelector('.home-contents-container')

  const subs = await sdb.watch(onbatch)

  // -------------------- Create components --------------------
  // Pass opts with wallet info for initial BTC view
  const headerOpts = { wallet: 'btc', amount: '0.9616' }
  const home_page_header_component = await home_page_header(subs[0], headerOpts)
  const action_buttons_component = await action_buttons(subs[1], home_protocol)
  const transaction_list_component = await transaction_list(subs[2])
  const total_wealth_component = await total_wealth(subs[3])
  const lightning_buttons_component = await lightning_buttons(subs[4], home_protocol)
  const wallet_button_component = await wallet_button(subs[5], home_protocol)
  const light_page_header_component = await light_page_header(subs[6])
  const light_tx_list_component = await light_tx_list(subs[7])

  // default view: BTC wallet components
  container.append(
    home_page_header_component,
    wallet_button_component,
    action_buttons_component,
    transaction_list_component,
    total_wealth_component
  )

  return el

  // -------------------- Protocol to handle child messages --------------------
  function home_protocol(sendToChild) {
    return async function (messageFromChild) {
      console.log('[child -> home_contents]:', messageFromChild)

      if (messageFromChild?.from === 'switch_account' && messageFromChild.type === 'switch') {
        const wallet = messageFromChild.data

        if (wallet === 'btc') {
          // Update header dynamically
          home_page_header_component._update?.({ wallet: 'btc', amount: '0.9616' })

          // Re-append BTC view components
          container.innerHTML = ''
          container.append(
            home_page_header_component,
            wallet_button_component,
            action_buttons_component,
            transaction_list_component,
            total_wealth_component
          )

          console.log('Switched to BTC wallet view')
        } else if (wallet === 'lightning') {
          // Update header dynamically
          home_page_header_component._update?.({ wallet: 'lightning', amount: '0.0246' })

          // Show Lightning buttons instead of BTC action buttons
          container.innerHTML = ''
          container.append(
            light_page_header_component,
            wallet_button_component,
            lightning_buttons_component,
            light_tx_list_component,
            total_wealth_component
          )

          console.log('Switched to Lightning wallet view')
        }
      }
    }
  }

  async function onbatch(batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(p => drive.get(p).then(f => f.raw)))
      const func = on[type] || fail
      await func(data, type)
    }
  }

  function inject(data) { style.textContent = data[0] }
  function ondata(data) {}
  function fail(data, type) { throw new Error('invalid message', { cause: { data, type } }) }
}

// ------------------ Fallback for STATE ------------------
function fallback_module() {
  return {
    api,
    _: {
      'home_page_header': { $: '' },
      'action_buttons': { $: '' },
      'transaction_list': { $: '' },
      'total_wealth': { $: '' },
      'lightning_buttons': { $: '' },
      'wallet_button': { $: '' },
      'light_page_header': { $: '' },
      'light_tx_list': { $: '' }
    }
  }

  function api(opts) {
    const home_page_header = {
      mapping: { style: 'style', data: 'data', icons: 'icons' },
      0: { wallet: 'btc', amount: '0.9616' }
    }
    const action_buttons = { mapping: { style: 'style', data: 'data' }, 0: { buttons: { wallet: 'btc' } } }
    const transaction_list = {
      mapping: { style: 'style', data: 'data' },
      0: { value: [
        { tid: 'Luis fedrick', ttime: '11:30 AM', tamount: '+0.02456', avatar: 'https://tse2.mm.bing.net/th/id/OIP.255ajP8y6dHwTTO8QbBzqwHaHa?rs=1&pid=ImgDetMain&o=7&rm=' },
        { tid: 'skdmf932ksdmf0234lsd', ttime: '02:15 PM', tamount: '+0.03271', avatar: '' },
        { tid: 'Mark Kevin', ttime: '03:45 PM', tamount: '-0.00421', avatar: 'https://tse4.mm.bing.net/th/id/OIP.VIRWK2jj8b2cHBaymZC5AgHaHa?w=800&h=800&rs=1&pid=ImgDetMain&o=7&rm=3' },
        { tid: 'yweuyiewe32eqw234lsd', ttime: '12:30 PM', tamount: '+0.00567', avatar: '' },
      ]}
    }
    const total_wealth = {
      mapping: { style: 'style', data: 'data', icons: 'icons' },
      0: { value: { total: 0.9862, usd: 1000, lightning: 0.02456, bitcoin: 0.96164 } }
    }
    const lightning_buttons = { mapping: { style: 'style', data: 'data' }, 0: { buttons: { wallet: 'lightning' } } }
    const wallet_button = { mapping: { style: 'style', data: 'data' }, 0: { buttons: { wallet: 'btc' } } }
    const light_page_header = {
      mapping: { style: 'style', data: 'data', icons: 'icons' },
      0: { wallet: 'lightning', amount: '0.0246' }
    }
    const light_tx_list = {
      mapping: { style: 'style', data: 'data' },
      0: { value: [
        { tid: 'skdmf932ksdmf0234lsd', ttime: '09:15 AM', tamount: '-0.00123', avatar: '' },
        { tid: 'yweuyiewe32eqw234lsd', ttime: '12:30 PM', tamount: '+0.00567', avatar: '' },
        { tid: 'skdmf932hiuo1h2oihio', ttime: '04:20 PM', tamount: '-0.00234', avatar: '' },
        { tid: 'yweuyiewe32eqw234lsd', ttime: '12:30 PM', tamount: '+0.00567', avatar: '' },
      ]}
    }

    return {
      drive: { 'style/': { 'home_contents.css': { '$ref': 'home_contents.css' } }, 'data/': { 'opts.json': { raw: opts } } },
      _: { home_page_header, action_buttons, transaction_list, total_wealth, lightning_buttons, wallet_button, light_page_header, light_tx_list }
    }
  }
}
