// send_to_contact/send_to_contact.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { sdb, get } = statedb(fallback_module)

module.exports = send_to_contact

async function send_to_contact (opts = {}) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb
  
  const on = {
    style: inject,
    data: ondata,
    icons: iconject,

  }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  shadow.innerHTML = `
    <div class="contact-row"></div>
    <style></style>
  `
  const style = shadow.querySelector('style')
  const row = shadow.querySelector('.contact-row')
  
  let dricons = []
  let selected = false

  await sdb.watch(onbatch)

  row.onclick = () => {
    selected = !selected
    update_tick()
    update_selection()
  }

  return el


  function update_tick() {
    const avatarWrapper = row.querySelector('.contact-avatar')
    if (!avatarWrapper) return

    const oldTick = avatarWrapper.querySelector('.green-tick')
    if (oldTick) oldTick.remove()

    if (selected && dricons[0]) {
      avatarWrapper.innerHTML += `
        <div class="green-tick">
          ${dricons[0] || ''}
        </div>
      `
    }
  }

  function update_selection() {
    if (selected) {
      row.classList.add('selected')
    } else {
      row.classList.remove('selected')
    }
  }

  function fail(data, type) {
    throw new Error('invalid message', { cause: { data, type } })
  }

  async function onbatch (batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(
        paths.map(path => drive.get(path).then(file => file.raw))
      )
      const func = on[type] || fail
      await func(data, type)
    }
  }

  function inject (data) {
    style.textContent = data[0]
  }

  async function ondata (data) {
    const { avatar, name } = data[0]

    row.innerHTML = `
      <div class="contact-left">
        <div class="contact-avatar">
          <img src="${avatar}" alt="avatar" />
        </div>
        <div class="contact-info">
          <div class="contact-name">${name}</div>
        </div>
      </div>
    `

    update_tick() 
    update_selection()
  }

   function iconject (data) {
    dricons = data
  }
}


function fallback_module () {
  return {
    api: fallback_instance,
  }
  function fallback_instance (opts) {
    if (!opts) opts = {}
    if (!opts.value) opts.value = {}

    return {
        drive: {
          'icons/': {
            'green-tick.svg': {
              '$ref': 'green-tick.svg'
            },
          },
          'style/':{
            'send_to_contact.css': {
              '$ref': 'send_to_contact.css'
            },
          },
          'data/': {
            'opts.json': {
              raw: opts || {}
            }
          }
        }
    }
  }
}
