// contact_row/contact_row.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { sdb, get } = statedb(fallback_module)

const chat_view = require('chat_view')

module.exports = contact_row

async function contact_row(opts = {}, protocol) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = { style: inject, data: ondata, icons: iconject }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  shadow.innerHTML = `
    <div class="contact-row"></div>
    <div class="chat-view"></div>

    <style></style>
  `
  const style = shadow.querySelector('style')
  const row = shadow.querySelector('.contact-row')
  const chat_view_container = shadow.querySelector('.chat-view')

  let dricons = []

  const sendToParent = msg => protocol?.(msg)

  const subs =  await sdb.watch(onbatch)

  const chat_view_comp = await chat_view(subs[0])
  
  chat_view_container.append(chat_view_comp)

  

  return el

  function fail(data, type) {
    throw new Error('invalid message', { cause: { data, type } })
  }

  async function onbatch(batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(path => drive.get(path).then(f => f.raw)))
      const func = on[type] || fail
      await func(data, type)
    }
  }

  function inject(data) { style.textContent = data[0] }

  async function ondata(data) {
    const { avatar, name, message, time, unread, online, lightining } = data[0] || {}

    row.innerHTML = `
      <div class="contact-left">
          <div class="contact-avatar">
            <img src="${avatar || ''}" alt="avatar" />
            ${online ? '<div class="online-dot"></div>' : ''}
          </div>
          <div class="contact-info">
            <div class="contact-name">${name || ''}</div>
            <div class="contact-message ${unread > 0 ? 'unread-message' : ''}">${message || ''}</div>       
        </div>
      </div>
      <div class="contact-right">
        <div class="contact-time">${time || ''}</div>
        <div class="icon-wrapper  ${!lightining ? 'no-lightning' : ''}">
          ${lightining && dricons[0] ? dricons[0] : ''}
          ${unread > 0 ? `<div class="unread-badge">${unread}</div>` : ''}
        </div>
      </div>` 

    // Notify parent with contact name
    sendToParent({ type: 'contact-name', name: name || '' })
  }

  function iconject(data) { dricons = data }
}

function fallback_module () {
return {
  api: fallback_instance,
  _:{
    chat_view: { $: ''},
  }
}
function fallback_instance (opts) {
  const chat_view = {
    mapping: { style: 'style', data: 'data', icons: 'icons' },
      0: opts.value,
  }
  return {
      drive: {
        'icons/': {
          'lightning.svg': {
            '$ref': 'lightning.svg'
          },
        },
        'style/':{
          'style.css':{
            '$ref': 'contact_row.css'
          }
        },
        'data/': {
          'opts.json': {
            raw: opts
          }
        }
      },
      _: {
        chat_view
      }
  }
}
}
