// create_invoice_confirmation/create_invoice_confirmation.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { sdb, get } = statedb(fallback_module)

const receipt_row = require('receipt_row')
const send_button = require('button') 
const send_invoice_modal = require('send_invoice_modal')

module.exports = create_invoice_confirmation

async function create_invoice_confirmation(opts = {}, { onClose } = {}) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = { style: inject, data: ondata, icons: iconject }

  let dricons = []
  let modalMounted = false

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  shadow.innerHTML = `
    <div class="invoice-card">
      <div class="invoice-header">
        <div class="title-container">
          <div class="invoice-title">Create Lightning Invoice</div>
          <div class="light-icon-small"></div>
        </div>
        <div class="close-icon"></div>
      </div>
      <div class="receipt-rows"></div>
      <div class="send-button"></div>
      <div class="send-modal-container hidden"></div>
      <style></style>
    </div>
  `

  const style = shadow.querySelector('style')
  const rows_el = shadow.querySelector('.receipt-rows')
  const send_button_container = shadow.querySelector('.send-button')
  const close_icon = shadow.querySelector('.close-icon')
  const modal_container = shadow.querySelector('.send-modal-container')

  // Close the invoice card
  close_icon.addEventListener('click', () => {
    if (onClose) onClose()
  })

  const subs = await sdb.watch(onbatch)

  // Render send button (subs[0])
  const button_component = await send_button(subs[0])
  send_button_container.append(button_component)

  // Render send_invoice_modal only when button is clicked (subs[1])
  button_component.addEventListener('click', async () => {
    modal_container.classList.remove('hidden')
    if (!modalMounted) {
      const modal_component = await send_invoice_modal(subs[1], {
        onClose: () => modal_container.classList.add('hidden')
      })
      modal_container.append(modal_component)
      modalMounted = true
    }
  })

  // Render receipt rows (subs[2+] if any)
  for (let i = 2; i < subs.length; i++) {
    const row_component = await receipt_row(subs[i])
    rows_el.append(row_component)
  }

  return el

  function fail(data, type) {
    throw new Error('invalid message', { cause: { data, type } })
  }

  async function onbatch(batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(p => drive.get(p).then(f => f.raw)))
      const fn = on[type] || fail
      await fn(data, type)
    }
  }

  function inject(data) {
    style.textContent = data[0]
  }

  function iconject(data) {
    dricons = data
    const light_icon = shadow.querySelectorAll('.light-icon-small')
    const close_icon = shadow.querySelector('.close-icon')
    light_icon.forEach(el => el.innerHTML = dricons[0])
    close_icon.innerHTML = dricons[1]
  }

  async function ondata(data) {}
}

function fallback_module() {
  return {
    api,
    _: {
      'button': { $: '' },
      'send_invoice_modal': { $: '' },
      'receipt_row': { $: '' },
    }
  }

  function api(opts) {
    const button = { mapping: { style: 'style', data: 'data', icons: 'icons' }, 0: { label: 'Send' } }
    const send_invoice_modal = { mapping: { style: 'style', data: 'data', icons: 'icons' }, 1: {
       value: [
            {
              avatar: "https://tse4.mm.bing.net/th/id/OIP.VIRWK2jj8b2cHBaymZC5AgHaHa?w=800&h=800&rs=1&pid=ImgDetMain&o=7&rm=3",
              name: 'Mark Kevin',
              message: 'Payment Received successfully',
              time: '3 hr',
              unread: 5,
              online: true,
              lightining: true
            },
            {
              avatar: "https://tse4.mm.bing.net/th/id/OIP.bdn3Kne-OZLwGM8Uoq5-7gHaHa?w=512&h=512&rs=1&pid=ImgDetMain&o=7&rm=3",
              name: 'David Clark',
              message: 'You have a new message from Mark',
              time: '1 hr',
              unread: 5,
              online: false,
              lightining: false
            },
            {
              avatar: "https://tse4.mm.bing.net/th/id/OIP.bdn3Kne-OZLwGM8Uoq5-7gHaHa?w=512&h=512&rs=1&pid=ImgDetMain&o=7&rm=3",
              name: 'David Clark',
              message: 'Received funds',
              time: '1 hr',
              unread: 0,
              online: true,
              lightining: true
            },
            {
              avatar: "https://tse4.mm.bing.net/th/id/OIP.7XLV6q-D_hA-GQh_eJu52AHaHa?rs=1&pid=ImgDetMain&o=7&rm=3",
              name: 'Sara Ahmed',
              message: 'Invoice sent',
              time: '2 hr',
              unread: 0,
              online: false,
              lightining: false
            }
          ] 
        }
      }

    const receipt_row = { mapping: { style: 'style', data: 'data', icons: 'icons' } }

    opts.value.slice(1).forEach((row, i) => {
      receipt_row[i + 2] = row
    })

    return {
      drive: {
        'icons/': { 'lightning.svg': { '$ref': 'lightning.svg' }, 'x.svg': { '$ref': 'x.svg' } },
        'style/': { 'create_invoice_confirmation.css': { '$ref': 'create_invoice_confirmation.css' } },
        'data/': { 'opts.json': { raw: opts } }
      },
      _: { button, send_invoice_modal, receipt_row}
    }
  }
}
