// create_invoice/create_invoice.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { sdb, get } = statedb(fallback_module)

const create_button = require('button')
const btc_input_card = require('btc_input_card')
// const input_field = require('input_field')

module.exports = create_invoice

async function create_invoice(opts = {}) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = {
    style: inject,
    data: ondata,
    icons: iconject,
  }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  let dricons = []

  shadow.innerHTML = `
    <div class="create-invoice-container">
      <div class="create-invoice-header">  
        <div class="title-container"> 
          <div class="create-invoice-header">Create Lightning Invoice</div>
          <div class="btc-icon"></div>
        </div>  
        <div class="x-icon"></div>
      </div>
      <div class="btc-input-card"></div>
      <div class="extra-input1"></div>
      <div class="extra-input2"></div>
      <div class="create_button"></div>
    </div>
    <style></style>
  `

  const style = shadow.querySelector('style')
  const create_button_component = shadow.querySelector('.create_button')
  const btc_input_card_component = shadow.querySelector('.btc-input-card')
  const input1_component = shadow.querySelector('.extra-input1')

  const subs = await sdb.watch(onbatch)

  const button_component = await create_button(subs[0])
  const btc_component = await btc_input_card(subs[1])
  // const input1 = await input_field(subs[2])

  create_button_component.append(button_component)
  btc_input_card_component.append(btc_component)
  // input1_component.append(input1)

  const closeBtn = shadow.querySelector('.x-icon')
  if (closeBtn) {
    closeBtn.onclick = () => {
      const dropdown = el.parentNode
      if (dropdown) {
        dropdown.classList.add('hidden')   
      }
    }
  }

  return el

  function fail(data, type) {
    throw new Error('invalid message', { cause: { data, type } })
  }

  async function onbatch(batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(path => drive.get(path).then(file => file.raw)))
      const func = on[type] || fail
      await func(data, type)
    }
  }

  function inject(data) {
    style.textContent = data[0]
  }

  function iconject (data) {
    dricons = data

    const btc_icon = shadow.querySelector('.btc-icon')
    const x_icon = shadow.querySelector('.x-icon')

    btc_icon.innerHTML = dricons[0]
    x_icon.innerHTML = dricons[1]
  }

  async function ondata(data) {
    
  }
}

function fallback_module() {
  return {
    api,
    _: {
      'button': { $: '' },
      'btc_input_card': { $: '' },
      // 'input_field': { $: '' },
    }
  }

  function api(opts) {
    
    const button = {
      mapping: {
        style: 'style',
        data: 'data',
      },
      0: {
        label: 'Create Invoice'
      }    
    }
    
    const btc_input_card = {
      mapping: {
        style: 'style',
        data: 'data',
      },
      1: {
        balance: 0.0024, 
      }  
    }

    // const input_field = {
    //   mapping: {
    //     style: 'style',
    //     data: 'data',
    //     icons: 'icons'
    //   },
    //   2: {
    //     header: 'Invoice Title',
    //     placeholder: 'Enter invoice title',
    //   }    
    // }


    return {
      drive: {
        'icons/': {
          'lightning.svg': {
            '$ref': 'lightning.svg'
          },
          'x.svg': {
            '$ref': 'x.svg'
          },
        },
        'style/': {
          'create_invoice.css': {
            '$ref': 'create_invoice.css'
          }
        },
        'data/': {
          'opts.json': {
            raw: opts
          }
        }
      },
      _: {
        button,
        btc_input_card,
        // input_field      
      }
    }
  }
}
