// pay_invoice/pay_invoice.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { sdb, get } = statedb(fallback_module)

const pay_button = require('button')
const invoice_input = require('input_field')

module.exports = pay_invoice

async function pay_invoice(opts = {}) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = {
    style: inject,
    data: ondata,
    icons: iconject,
  }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  let dricons = []

  shadow.innerHTML = `
    <div class="pay-invoice-container">
      <div class="pay-invoice-header">  
        <div class="title-container"> 
          <div class="header-title">Pay Lightning Invoice</div>
          <div class="lightning-icon"></div>
        </div>  
        <div class="x-icon"></div>
      </div>
      <div class="invoice-input"></div>
      <div class="pay_button"></div>
    </div>
    <style></style>
  `

  const style = shadow.querySelector('style')
  const pay_button_component = shadow.querySelector('.pay_button')
  const invoice_input_component = shadow.querySelector('.invoice-input')

  const subs = await sdb.watch(onbatch)

  const button_component = await pay_button(subs[0])
  const invoice_component = await invoice_input(subs[1])

  pay_button_component.append(button_component)
  invoice_input_component.append(invoice_component)

  const closeBtn = shadow.querySelector('.x-icon')
  if (closeBtn) {
    closeBtn.onclick = () => {
      const dropdown = el.parentNode
      if (dropdown) {
        dropdown.classList.add('hidden')   
      }
    }
  }

  return el

  function fail(data, type) {
    throw new Error('invalid message', { cause: { data, type } })
  }

  async function onbatch(batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(path => drive.get(path).then(file => file.raw)))
      const func = on[type] || fail
      await func(data, type)
    }
  }

  function inject(data) {
    style.textContent = data[0]
  }

  function iconject (data) {
    dricons = data

    const lightning_icon = shadow.querySelector('.lightning-icon')
    const x_icon = shadow.querySelector('.x-icon')

    lightning_icon.innerHTML = dricons[0]
    x_icon.innerHTML = dricons[1]
  }

  async function ondata(data) {}
}

function fallback_module() {
  return {
    api,
    _: {
      'button': { $: '' },
      'input_field': { $: '' },
    }
  }

  function api(opts) {
    const button = {
      mapping: {
        style: 'style',
        data: 'data',
      },
      0: {
        label: 'Pay Invoice'
      }    
    }
    
    const input_field = {
      mapping: {
        style: 'style',
        data: 'data',
        icons: 'icons'
      },
      1: {
        header: 'Lightning Invoice',
        placeholder: 'Tap to paste your lightning invoice',
        address: 'lnbc1u1p0exampleinvoicehere1234567890'  // demo value
      }    
    }

    return {
      drive: {
        'icons/': {
          'lightning.svg': {
            '$ref': 'lightning.svg'
          },
          'x.svg': {
            '$ref': 'x.svg'
          },
        },
        'style/': {
          'pay_invoice.css': {
            '$ref': 'pay_invoice.css'
          }
        },
        'data/': {
          'opts.json': {
            raw: opts
          }
        }
      },
      _: {
        button,
        input_field
      }
    }
  }
}
