// pay_invoice/pay_invoice.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { sdb, get } = statedb(fallback_module)

const pay_button = require('button')
const invoice_input = require('input_field')
const pay_invoice_confirmation = require('pay_invoice_confirmation')

module.exports = pay_invoice

async function pay_invoice(opts = {}) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = {
    style: inject,
    data: ondata,
    icons: iconject,
  }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  let dricons = []

  shadow.innerHTML = `
    <div class="pay-invoice-container">
      <div class="pay-invoice-header">  
        <div class="title-container"> 
          <div class="header-title">Pay Lightning Invoice</div>
          <div class="lightning-icon"></div>
        </div>  
        <div class="x-icon"></div>
      </div>
      <div class="invoice-input"></div>
      <div class="pay_button"></div>
    </div>
    <div class="confirmation-container hidden"></div>
    <style></style>
  `

  const style = shadow.querySelector('style')
  const button_container = shadow.querySelector('.pay_button')
  const invoice_input_component = shadow.querySelector('.invoice-input')
  const confirmation_container = shadow.querySelector('.confirmation-container')

  const subs = await sdb.watch(onbatch)

  const button_component = await pay_button(subs[0])
  const invoice_component = await invoice_input(subs[1])

  button_container.append(button_component)
  invoice_input_component.append(invoice_component)

  let invoice_mounted = false 

  const closeBtn = shadow.querySelector('.x-icon')
  if (closeBtn) {
    closeBtn.onclick = () => {
      const dropdown = el.parentNode
      if (dropdown) {
        dropdown.classList.add('hidden')   
      }
    }
  }

  button_container.addEventListener('click', async () => {
    confirmation_container.classList.remove('hidden')

      if (!invoice_mounted) {
      const invoice = await pay_invoice_confirmation(subs[2], {
        onClose: () => confirmation_container.classList.add('hidden')
      })
      confirmation_container.append(invoice)
      invoice_mounted = true
    }
  })

  return el

  function fail(data, type) {
    throw new Error('invalid message', { cause: { data, type } })
  }

  async function onbatch(batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(path => drive.get(path).then(file => file.raw)))
      const func = on[type] || fail
      await func(data, type)
    }
  }

  function inject(data) {
    style.textContent = data[0]
  }

  function iconject (data) {
    dricons = data

    const lightning_icon = shadow.querySelector('.lightning-icon')
    const x_icon = shadow.querySelector('.x-icon')

    lightning_icon.innerHTML = dricons[0]
    x_icon.innerHTML = dricons[1]
  }

  async function ondata(data) {}
}

function fallback_module() {
  return {
    api,
    _: {
      'button': { $: '' },
      'input_field': { $: '' },
      'pay_invoice_confirmation': { $: '' },
    }
  }

  function api(opts) {
    if (!opts) opts = {}
    if (!opts.value) opts.value = {}
 
    const button = {
      mapping: {
        style: 'style',
        data: 'data',
      },
      0: {
        label: 'Pay Invoice'
      }    
    }
    
    const input_field = {
      mapping: {
        style: 'style',
        data: 'data',
        icons: 'icons'
      },
      1: {
        header: 'Lightning Invoice',
        placeholder: 'Tap to paste your lightning invoice',
        address: 'lnbc1u1p0exampleinvoicehere1234567890'  // demo value
      }    
    }

    const pay_invoice_confirmation = {
      mapping: {
        style: 'style',
        data: 'data',
        icons: 'icons'
      },
      2: {
        value: [
              { label: "Amount", value: "0.0030 BTC", convert: true },
              { label: "Fee", value: "0.0001 BTC", convert: true },
              { label: "Recipient Address", value: "7RwmbHfn...455p" },
              { label: "Processing time", value: "< 5 minutes" },
              { label: "Total (inc. fee)", value: "0.0031 BTC ",  icon: "lightning.svg", convert: true }
            ]
      },  
    }

    return {
      drive: {
        'icons/': {
          'lightning.svg': {
            '$ref': 'lightning.svg'
          },
          'x.svg': {
            '$ref': 'x.svg'
          },
        },
        'style/': {
          'pay_invoice.css': {
            '$ref': 'pay_invoice.css'
          }
        },
        'data/': {
          'opts.json': {
            raw: opts || {}
          }
        }
      },
      _: {
        button,
        input_field,
        pay_invoice_confirmation
      }
    }
  }
}
