// gen_invite_code/gen_invite_code.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { sdb, get } = statedb(fallback_module)

const invoice_input = require('input_field')

module.exports = gen_invite_code

async function gen_invite_code(opts = {}, { onClose } = {}) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = {
    style: inject,
    data: ondata,
    icons: iconject,
  }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  let dricons = []

  shadow.innerHTML = `
    <div class="pay-invoice-container">
      <div class="pay-invoice-header">  
        <div class="title-container"> 
          <div class="header-title">Generate Invite Code</div>
        </div>  
        <div class="x-icon"></div>
      </div>
      <div class="invoice-input"></div>
    </div>
    <style></style>
  `

  const style = shadow.querySelector('style')
  const invoice_input_component = shadow.querySelector('.invoice-input')

  const subs = await sdb.watch(onbatch)

  const invoice_component = await invoice_input(subs[0])

  invoice_input_component.append(invoice_component)

  const closeBtn = shadow.querySelector('.x-icon')
  if (closeBtn) {
    closeBtn.onclick = () => {
      if (onClose) onClose()
    }
  }


  return el

  function fail(data, type) {
    throw new Error('invalid message', { cause: { data, type } })
  }

  async function onbatch(batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(path => drive.get(path).then(file => file.raw)))
      const func = on[type] || fail
      await func(data, type)
    }
  }

  function inject(data) {
    style.textContent = data[0]
  }

  function iconject (data) {
    dricons = data

    const x_icon = shadow.querySelector('.x-icon')

    x_icon.innerHTML = dricons[1]
  }

  async function ondata(data) {}
}

function fallback_module() {
  return {
    api,
    _: {
      'input_field': { $: '' },
    }
  }

  function api(opts) {
    if (!opts) opts = {}
    if (!opts.value) opts.value = {}
    
    const input_field = {
      mapping: {
        style: 'style',
        data: 'data',
        icons: 'icons'
      },
      0: {
        header: 'Your Invite Code',
        placeholder: 'Paste your code here',
        address: 'lnbc1u1p0exampleinvoicehere1234567890',
        icon: 'apple'  // demo value
      }    
    }

    return {
      drive: {
        'icons/': {
          'lightning.svg': {
            '$ref': 'lightning.svg'
          },
          'x.svg': {
            '$ref': 'x.svg'
          },
        },
        'style/': {
          'gen_invite_code.css': {
            '$ref': 'gen_invite_code.css'
          }
        },
        'data/': {
          'opts.json': {
            raw: opts || {}
          }
        }
      },
      _: {
        input_field
      }
    }
  }
}
