// switch_send/switch_send.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { sdb, get } = statedb(fallback_module)

const send_btc = require('send_btc')
const pending_request = require('pending_request')

module.exports = switch_send

async function switch_send(opts = {}, protocol) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = {
    style: inject,
    data: ondata,
    icons: iconject,
  }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  let dricons = []
  let showSendBtc = false

  shadow.innerHTML = `
    <div class="parent-wrapper">
      <div class="switch-account-container">
        <div class="account-content"></div>
      </div>
    </div>
    <div class="send-btc hidden"></div>
    <div class="send-light hidden"></div>

    <style></style>
  `

  const content = shadow.querySelector('.account-content')
  const style = shadow.querySelector('style')
  const send_btc_el = shadow.querySelector('.send-btc')
  const send_light_el = shadow.querySelector('.send-light')

  const subs = await sdb.watch(onbatch)

  let btc_mounted = false
  let light_mounted = false

  return el

  async function onbatch(batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(p => drive.get(p).then(f => f.raw)))
      const fn = on[type] || fail
      await fn(data, type)
    }
  }

  function inject(data) {
    style.textContent = data[0]
  }

  async function ondata(data) {
    const { btc, lightning } = data[0] || {}

    if (showSendBtc) return

    content.innerHTML = `
      <div class="account-container btc-container">
        <div class="btc-icon">${dricons[1] || '₿'} BTC</div>
        <div class="btc-amount">${btc ? parseFloat(btc).toFixed(4) : '0.0000'}</div>
      </div>

      <div class="account-container lightning-container">
        <div class="lightning-icon">${dricons[2] || '⚡'} Lightning</div>
        <div class="lightning-amount">${lightning ? parseFloat(lightning).toFixed(4) : '0.0000'}</div>
      </div>
    `
    const btc_container = content.querySelector('.btc-container')
    const lightning_container = content.querySelector('.lightning-container')
 
    btc_container.addEventListener('click', async () => {
      send_btc_el.classList.remove('hidden')

      if (!btc_mounted) {
        const send_btc_component = await send_btc(subs[0], {
          onClose: () => send_btc_el.classList.add('hidden')
        })
        send_btc_el.append(send_btc_component)
        btc_mounted = true
      }
    })
  
    lightning_container.addEventListener('click', async () => {
      send_light_el.classList.remove('hidden')

      if (!light_mounted) {
        const send_light = await pending_request(subs[1], {
          onClose: () => send_light_el.classList.add('hidden')
        })
        send_light_el.append(send_light)
        light_mounted = true
      }
    })


  }

  function iconject(data) {
    dricons = data
  }

  function fail(data, type) {
    throw new Error("invalid message", { cause: { data, type } })
  }
}



// ================== fallback ==================
function fallback_module() {
  return {
    api: fallback_instance,
    _: {
      'send_btc': { $: '' },
      'pending_request': { $: '' }
    }
  }

  function fallback_instance(opts) {
    if (!opts) opts = {}
    if (!opts.value) opts.value = {}

    const send_btc = {
      mapping: { style: 'style', data: 'data', icons: 'icons' },
      0: {}
    }

    const pending_request = {
      mapping: { style: 'style', data: 'data', icons: 'icons' },
      1: {}
    }

    return {
      drive: {
        'icons/': {
          'x.svg': { '$ref': 'x.svg' },
          'btc.svg': { '$ref': 'btc.svg' },
          'lightning.svg': { '$ref': 'lightning.svg' }
        },
        'style/': {
          'style.css': { '$ref': 'switch_send.css' }
        },
        'data/': {
          'opts.json': { raw: opts || {} }
        }
      },
      _: { send_btc, pending_request }
    }
  }
}
