// switch_send/switch_send.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { sdb, get } = statedb(fallback_module)

const send_btc = require('send_btc')
const pending_request = require('pending_request')

module.exports = switch_send

async function switch_send(opts = {}, protocol) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = {
    style: inject,
    data: ondata,
    icons: iconject,
  }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  let dricons = []
  let showSendBtc = false
  let send_btc_comp = null

  shadow.innerHTML = `
    <div class="parent-wrapper">
      <div class="pop-up"></div>

      <div class="switch-account-container">
        <div class="account-content"></div>
      </div>
    </div>

    <style></style>
  `

  const popup = shadow.querySelector('.pop-up')
  const container = shadow.querySelector('.switch-account-container')
  const content = shadow.querySelector('.account-content')
  const style = shadow.querySelector('style')

  popup.style.display = "none"

  // ONLY ONE SUBS EVER
  const subs = await sdb.watch(onbatch)

  // Build send_btc ONE time
  send_btc_comp = await send_btc(subs[0])
  send_btc_comp.classList.add("send-btc")

  pending_req_comp = await pending_request(subs[1])
  pending_req_comp.classList.add("pending-requests")

  return el

  async function onbatch(batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(p => drive.get(p).then(f => f.raw)))
      const fn = on[type] || fail
      await fn(data, type)
    }
  }

  function inject(data) {
    style.textContent = data[0]
  }

  async function ondata(data) {
    const { btc, lightning } = data[0] || {}

    // If popup is open (send_btc visible), do NOT overwrite BTC/Lightning UI
    if (showSendBtc) return

    content.innerHTML = `
      <div class="account-container btc-container">
        <div class="btc-icon">${dricons[1] || '₿'} BTC</div>
        <div class="btc-amount">${btc ? parseFloat(btc).toFixed(4) : '0.0000'}</div>
      </div>

      <div class="account-container lightning-container">
        <div class="lightning-icon">${dricons[2] || '⚡'} Lightning</div>
        <div class="lightning-amount">${lightning ? parseFloat(lightning).toFixed(4) : '0.0000'}</div>
      </div>
    `

    // BTC CLICK → toggle popup visibility
    content.querySelector('.btc-container').onclick = () => {
      toggle_send_btc()
    }

    content.querySelector('.lightning-container').onclick = () => {
      togglet_pending_req()
    }
  }

  // Toggle open/close of send_btc
  function toggle_send_btc() {
    showSendBtc = !showSendBtc
    popup.classList.remove('hidden')

    if (showSendBtc) {
      popup.innerHTML = ""
      popup.appendChild(send_btc_comp)
      popup.style.display = "flex"
    } else {
      popup.style.display = "none"
      popup.innerHTML = ""
    }
  }

  function togglet_pending_req() {
    showSendBtc = !showSendBtc
    popup.classList.remove('hidden')

    if (showSendBtc) {
      popup.innerHTML = ""
      popup.appendChild(pending_req_comp)
      popup.style.display = "flex"
    } else {
      popup.style.display = "none"
      popup.innerHTML = ""
    }
  }


  function iconject(data) {
    dricons = data
  }

  function fail(data, type) {
    throw new Error("invalid message", { cause: { data, type } })
  }
}



// ================== fallback ==================
function fallback_module() {
  return {
    api: fallback_instance,
    _: {
      'send_btc': { $: '' },
      'pending_request': { $: '' }
    }
  }

  function fallback_instance(opts) {
    const send_btc = {
      mapping: { style: 'style', data: 'data', icons: 'icons' },
      0: {}
    }

    const pending_request = {
      mapping: { style: 'style', data: 'data', icons: 'icons' },
      1: {}
    }

    return {
      drive: {
        'icons/': {
          'x.svg': { '$ref': 'x.svg' },
          'btc.svg': { '$ref': 'btc.svg' },
          'lightning.svg': { '$ref': 'lightning.svg' }
        },
        'style/': {
          'style.css': { '$ref': 'switch_send.css' }
        },
        'data/': {
          'opts.json': { raw: opts }
        }
      },
      _: { send_btc, pending_request }
    }
  }
}
