// light_tx_list/light_tx_list.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { sdb, get } = statedb(fallback_module)

const light_tx_row = require('light_tx_row')
const light_tx_history = require('light_tx_history') 

module.exports = light_tx_list


async function light_tx_list(opts = {}) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = { style: inject, data: ondata }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'open' })

  shadow.innerHTML = `
    <div class="transaction-list-container">
      <div class="transaction-list-header">
        <div class="transaction-list-title">Transactions</div>
        <div class="transaction-list-see-all">See all</div>
      </div>
      <div class="transaction-list-body"></div>
    </div>

    <div class="transaction-overlay hidden">
      <div class="transaction-history-body"></div>
    </div>

    <style></style>
  `

  const style = shadow.querySelector('style')
  const listBody = shadow.querySelector('.transaction-list-body')
  const seeAllBtn = shadow.querySelector('.transaction-list-see-all')
  const overlay = shadow.querySelector('.transaction-overlay')
  const historyBody = shadow.querySelector('.transaction-history-body')

  let historyMounted = false

  const subs = await sdb.watch(onbatch)

  // render small list
  subs.slice(1, 5).forEach(async sub => {
    listBody.append(await light_tx_row(sub))
  })

  // open overlay
  seeAllBtn.addEventListener('click', async () => {
    overlay.classList.remove('hidden')

    if (!historyMounted) {
      const history = await light_tx_history(subs[0], {
        onClose: () => overlay.classList.add('hidden')
      })
      historyBody.append(history)
      historyMounted = true
    }
  })

  return el

  function fail(data, type) {
    throw new Error('invalid message', { cause: { data, type } })
  }

  async function onbatch(batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(
        paths.map(p => drive.get(p).then(f => f.raw))
      )
      const fn = on[type] || fail
      await fn(data)
    }
  }

  function inject(data) {
    style.textContent = data[0]
  }

  function ondata() {}
}

function fallback_module () {
  return {
    api,
    _: {
      'light_tx_history':{ $: ''},
      'light_tx_row':{ $: '' },
    } 
  }
  function api(opts){
    if (!opts) opts = {}
    if (!opts.value) opts.value = {}

    const light_tx_history = {
      mapping: {
        style: 'style',
        data: 'data'
      },
      0: {
        value:[
        { tid: 'skdmf932ksdmf0234lsd', ttime: '09:15 AM', tamount: '-0.00123', avatar: '' },
        { tid: 'yweuyiewe32eqw234lsd', ttime: '12:30 PM', tamount: '+0.00567', avatar: '' },
        { tid: 'skdmf932hiuo1h2oihio', ttime: '04:20 PM', tamount: '-0.00234', avatar: '' },
        { tid: 'yweuyiewe32eqw234lsd', ttime: '12:30 PM', tamount: '+0.00567', avatar: '' },
        { tid: 'skdmf932ksdmf0234lsd', ttime: '09:15 AM', tamount: '-0.00123', avatar: '' },
        { tid: 'yweuyiewe32eqw234lsd', ttime: '12:30 PM', tamount: '+0.00567', avatar: '' },
        { tid: 'skdmf932hiuo1h2oihio', ttime: '04:20 PM', tamount: '-0.00234', avatar: '' },
        { tid: 'yweuyiewe32eqw234lsd', ttime: '12:30 PM', tamount: '+0.00567', avatar: '' },
        ]
      }
    }


    const light_tx_row = {
      mapping: {
        style: 'style',
        data: 'data'
      }
    }
    opts.value.forEach((transaction, index) => {
      light_tx_row[index] = transaction
    })
    return {
      drive: {
        'style/': {
          'light_tx_list.css':{
            '$ref': 'light_tx_list.css'
          }
        },
        'data/': {
          'opts.json':{
            raw: opts || {}
          }
        }
      },
      _:{
        light_tx_history,
        light_tx_row
      }
    }
  }
}
