// btc_nodes/btc_nodes.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { sdb, get } = statedb(fallback_module)

module.exports = btc_nodes

async function btc_nodes (opts = {}) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = {
    style: inject,
    data: ondata,
    icons: iconject,
  }

  let dricons = []

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  shadow.innerHTML = `
    <div class="nodes-card"></div>
    <style></style>
  `
  const style = shadow.querySelector('style')
  const container = shadow.querySelector('.nodes-card')

  await sdb.watch(onbatch)

  return el

  function fail(data, type) {
    throw new Error('invalid message', { cause: { data, type } })
  }

  async function onbatch (batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(
        paths.map(path => drive.get(path).then(file => file.raw))
      )
      const func = on[type] || fail
      await func(data, type)
    }
  }

  function inject (data) {
    style.textContent = data[0]
  }

   


function iconject(data) {
  dricons = data  
}

async function ondata (data) {
  container.innerHTML = `
    <div class="header">
      <div class="dropdown">
        <button class="dropdown-btn">Bitcoin Nodes <span class="arrow"></span></button>
      </div>        
      <button class="close-btn">✕</button>
    </div>
    <div class="node-btn-wrap">
      <button class="node-btn">Node</button>
    </div>
    <div class="tab-indicator">
      <span class="dot active"></span>
      <span class="dot"></span>
    </div>
    <div class="logs-wrap">
      <label>Logs</label>
      <textarea class="logs-textarea"></textarea>
    </div>
  `

  // now insert icon if we already have it
  if (dricons.length) {
    const arrow_icons = shadow.querySelector('.arrow')
    if (arrow_icons) {
      arrow_icons.innerHTML = dricons[0]
    }
  }

  const closeBtn = container.querySelector('.close-btn')
  closeBtn.onclick = () => { container.innerHTML = "" }
}

}

function fallback_module () {
  return {
    api: fallback_instance,
  }
  function fallback_instance (opts) {
    return {
      drive: {
        'icons/': {
          'arrow-down.svg': { '$ref': 'arrow-down.svg' },
        },
        'style/': {
          'style.css': { '$ref': 'btc_nodes.css' },
        },
        'data/': {
          'opts.json': {
            raw: opts
          },
        }
      }
    }
  }
}
