// contacts_list/contacts_list.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { sdb, get } = statedb(fallback_module)

const contact_row = require('contact_row')

module.exports = contacts_list

async function contacts_list(opts = {}) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = {
    style: inject,
    data: ondata
  }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  shadow.innerHTML = `
    <div class="component-label">Contacts List</div>
    <div class="contact-list-container">
      <div class="contact-list-header">Contacts</div>
      <div class="search-bar">
        <input
          type="text"
          class="search-input"
          placeholder="Search"
          style="border: none; outline: none; font-size: 14px; background: transparent; flex: 1;"
        />
        <div class="search-icon">
          <svg width="22" height="22" viewBox="0 0 24 24" fill="none" stroke="#555" stroke-width="2" stroke-linecap="round" stroke-linejoin="round">
            <circle cx="11" cy="11" r="8" />
            <line x1="21" y1="21" x2="16.65" y2="16.65" />
          </svg>
        </div>
      </div>
    </div>
    <style></style>
  `

  const style = shadow.querySelector('style')
  const contact_list_container = shadow.querySelector('.contact-list-container')


  const subs = await sdb.watch(onbatch)
  
  subs.forEach(async sub => {
    contact_list_container.append(await contact_row(sub))
  })
  
  return el

  function fail(data, type) {
    throw new Error('invalid message', { cause: { data, type } })
  }

  async function onbatch(batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(path => drive.get(path).then(file => file.raw)))
      const func = on[type] || fail
      await func(data, type)
    }
  }

  function inject(data) {
    style.textContent = data[0]
  }

  async function ondata(data) {
  }

}

function fallback_module() {
  return {
    api,
    _: {
      'contact_row': {
        $: ''
      }
    }
  }
  function api(opts) {
    const contact_row = {
      mapping: {
        style: 'style',
        data: 'data'
      }
    }
    opts.value.forEach((contact, index) => {
      contact_row[index] = contact
    })
    return {
      drive: {
        'style/': {
          'contacts_list.css': {
            '$ref': 'contacts_list.css'
          }
        },
        'data/': {
          'opts.json': {
            raw: opts
          }
        }
      },
      _: {
        contact_row
      }
    }
  }
}
