// send_to/send_to.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { sdb, get } = statedb(fallback_module)

const search_bar = require('search_bar') 
const square_button = require('square_button') 
const send_to_contact = require('send_to_contact')
const share_invoice_via = require('share_invoice_via')

module.exports = send_to

async function send_to(opts = {}) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = {
    style: inject,
    data: ondata
  }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  shadow.innerHTML = `
    <div class="send-to-container"> 
      <div class="contact-list-container">
        <div class="contact-list-header">Send to</div>
        <div class="top-bar"></div>
      </div>
      <div class="contacts-container"></div>
      <div class="invoice-share-section"></div>
    </div>
    <style></style>
  `

  const style = shadow.querySelector('style')
  const contacts_container = shadow.querySelector('.contacts-container')
  const top_bar = shadow.querySelector('.top-bar')
  const invoice_section = shadow.querySelector('.invoice-share-section')

  const subs = await sdb.watch(onbatch)
  
  if (subs.length > 0) {
    const search = await search_bar(subs[0])
    const button = await square_button(subs[1])
    const invoice_share = await share_invoice_via(subs[2])
    top_bar.append(search)
    top_bar.append(button)  
    invoice_section.append(invoice_share)
  }

  for (let i = 3; i < subs.length; i++) {
    const contact = await send_to_contact(subs[i]) 
    contacts_container.append(contact)
  }

  return el

  function fail(data, type) {
    throw new Error('invalid message', { cause: { data, type } })
  }

  async function onbatch(batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(path => drive.get(path).then(file => file.raw)))
      const func = on[type] || fail
      await func(data, type)
    }
  }

  function inject(data) {
    style.textContent = data[0]
  }

  async function ondata(data) {
  }
}

function fallback_module() {
  return {
    api,
    _: {
      'search_bar': { $: '' },
      'send_to_contact': { $: '' },
      'share_invoice_via': { $: ''},
      'square_button': { $: '' },
    }
  }

  function api(opts) {
    const search_bar = {
      mapping: {
        style: 'style',
        data: 'data',
        icons: 'icons'
      },
      0: {
      }
    }

    const square_button = {
      mapping: {
        style: 'style',
        data: 'data',
        icons: 'icons'
      },
      1: {
      }
    }

    const share_invoice_via = {
      mapping: {
        style: 'style',
        data: 'data',
        icons: 'icons'
      },
      2: {
      }
    }

    const send_to_contact = {
      mapping: {
        style: 'style',
        data: 'data',
        icons: 'icons'

      }
    }

    opts.value.forEach((contact, index) => {
      send_to_contact[index + 3] = contact
    })

    return {
      drive: {
        'style/': {
          'send_to.css': {
            '$ref': 'send_to.css'
          }
        },
        'data/': {
          'opts.json': {
            raw: opts
          }
        }
      },
      _: {
        search_bar,
        square_button,
        share_invoice_via,
        send_to_contact
      }
    }
  }
}
