// btc_req_msg/btc_req_msg.js
const STATE = require('STATE')
const state_db = STATE(__filename)
const { sdb, get } = state_db(fallback_module)

module.exports = btc_req_msg

async function btc_req_msg(opts = {}) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = {
    style: inject_style,
    data: render_data,
    icons: inject_icons,
  }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })
  shadow.innerHTML = `
    <div class="btc_req_wrapper"></div>
    <style></style>
  `

  const style = shadow.querySelector('style')
  const wrapper = shadow.querySelector('.btc_req_wrapper')
  let dricons = []

  await sdb.watch(on_batch)
  return el

  function throw_error(data, type) {
    throw new Error('invalid message', { cause: { data, type } })
  }

  async function on_batch(batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(
        paths.map(path => drive.get(path).then(file => file.raw))
      )
      const fn = on[type] || throw_error
      await fn(data, type)
    }
  }

  function inject_style(data) {
    style.textContent = data[0]
  }

  function inject_icons(data) {
    dricons = data
  }

  function render_data(data) {
    const { avatar, name, amount, date, status, is_me } = data[0]

    const now = new Date()
    const expire_date = new Date(date)    
    let final_status = status

    // auto-change to expired if time has passed
    if (expire_date < now) {
      final_status = 'expired'
    }

    let border_color = '#DDB473'
    let btn_color = '#F7931A'
    let bg_color = '#FFFCE7'
    let status_element = ''

    if (final_status === 'send') {
      status_element = `<button class="status_btn" style="background:${btn_color};">Send</button>`
    } else if (final_status === 'paid') {
      border_color = '#87DD73'
      bg_color = '#ECFFE7'
      btn_color = '#4CAF50'
      status_element = `<button class="status_btn" style="background:${btn_color};">Paid √</button>`
    } else if (final_status === 'expired') {
      border_color = '#C83535'
      bg_color = '#FFE7E7'
      btn_color = '#C83535'
      status_element = `<button class="status_btn" style="background:${btn_color};">Expired</button>`
    }

    wrapper.innerHTML = `
      <div class="btc_row ${is_me ? 'self' : ''}">
        ${!is_me ? `<img class="avatar_outside" src="${avatar}" />` : ''}
        <div class="btc_box" style="border:2px solid ${border_color}; background:${bg_color}">
          <div class="btc_top">
            <span>
              <span class="name_text">${name}</span>
              <span class="req_text"> requests </span>
              <span class="amount_text">BTC ${amount}</span>
              <span class="btc_icon">${dricons[0] || '₿'}</span>
            </span>
          </div>
          <div class="btc_bottom">
            <span class="btc_expire">Expire at ${date}</span>
            ${status_element}
          </div>
        </div>
      </div>
    `
  }
}

function fallback_module() {
  return {
    api: fallback_instance,
  }

  function fallback_instance(opts) {
    return {
      drive: {
        'icons/': {
          'btc.svg': {
            '$ref': 'btc.svg' 
          }
        },
        'style/': {
          'btc_req_msg.css': {
            '$ref': 'btc_req_msg.css'
          }
        },
        'data/': {
          'opts.json': { raw: opts }
        }
      }
    }
  }
}
