// btc_req_msg/single_one.js
const STATE = require('STATE')
const state_db = STATE(__filename)
const { sdb, get } = state_db(fallback_module)

module.exports = btc_req_msg

async function btc_req_msg(opts = {}) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = {
    style: inject_style,
    data: render_data,
    icons: inject_icons,
  }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })
  shadow.innerHTML = `
    <div class="btc_req_wrapper"></div>
    <style></style>
  `

  const style = shadow.querySelector('style')
  const wrapper = shadow.querySelector('.btc_req_wrapper')
  let dricons = []

  await sdb.watch(on_batch)
  return el

  function throw_error(data, type) {
    throw new Error('invalid message', { cause: { data, type } })
  }

  async function on_batch(batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(
        paths.map(path => drive.get(path).then(file => file.raw))
      )
      const fn = on[type] || throw_error
      await fn(data, type)
    }
  }

  function inject_style(data) {
    style.textContent = data[0]
  }

  function inject_icons(data) {
    dricons = data
  }

 function render_data(data) {
    const { avatar, name, amount, date, status, is_me } = data[0]

    // default colors
    let border_color = '#DDB473'
    let btn_color = '#F7931A'
    let bg_color = '#FFFCE7'
    let status_element = ''

    if (status === 'send') {
      status_element = `<button class="status_btn" style="background:${btn_color};">${'Send'}</button>`
    } else if (status === 'paid') {
      border_color = '#87DD73'
      bg_color = '#ECFFE7'
      btn_color = '#4CAF50'
      status_element = `<button class="status_btn" style="background:${btn_color};">${'Paid √'}</button>`
    } else if (status === 'expired') {
      border_color = '#C83535'
      bg_color = '#FFE7E7'
      btn_color = '#C83535'
      status_element = `<button class="status_btn" style="background:${btn_color};">${'Expired'}</button>`
    }

    wrapper.innerHTML = `
      <div class="btc_row ${is_me ? 'self' : ''}">
        ${!is_me ? `<img class="avatar_outside" src="${avatar}" />` : ''}
        <div class="btc_box" style="border:2px solid ${border_color}; background:${bg_color}">
          <div class="btc_top">
            <span>
              <span class="name_text">${name}</span>
              <span class="req_text"> requests </span>
              <span class="amount_text">BTC ${amount} </span>
              <span class="btc_icon">${dricons[0] || '₿'}</span>
            </span>
          </div>
          <div class="btc_bottom">
            <span class="btc_expire">Expire at ${date}</span>
            ${status_element}
          </div>
        </div>
      </div>
    `
  }
}

function fallback_module() {
  return {
    api: fallback_instance,
  }

  function fallback_instance(opts) {
    return {
      drive: {
         'icons/': {
            'btc.svg': {
              '$ref': 'btc.svg'
            }
          },
        'style/': {
          'style.css': {
            raw: `
            .btc_row {
              display: flex;
              align-items: flex-start;
              gap: 12px;
              margin: 12px 0;
            }

            .btc_row.self {
              justify-content: flex-end;
            }

            .avatar_outside {
              width: 60px;
              height: 60px;
              border-radius: 50%;
              object-fit: cover;
            }

            .btc_box {
              border-radius: 14px;
              padding: 14px;
              min-width: 280px;
              border-width: 2px !important;
            }

            .btc_top {
              display: flex;
              flex-direction: column;
              gap: 3px;
            }

            .btc_bottom {
              display: flex;
              justify-content: space-between;
              align-items: center;
              margin-top: 6px; 
            }

            .name_text {
              font-size: 16px;
              font-weight: 600;
              color: #222;
            }


            .req_text {
              font-size: 15px;
              color: #777; 
            }

            .amount_text {
              font-size: 16px;
              font-weight: 700;
              color: #000; /* BTC amount stays black */
            }

            .btc_icon {
              margin-left: 4px;
            }

            .btc_expire {
              color: #555;
              font-weight: 500;
              font-size: 11px;
            }

            .send_btn {
              padding: 6px 16px;
              border-radius: 999px;
              font-size: 13px; 
              font-weight: 600;
              cursor: pointer;
              border: none;
              color: #fff;
            }

            .status_text {
              font-size: 12px;
              font-weight: 600;
            }
            .status_btn {
              padding: 6px 16px;
              border-radius: 999px;
              font-size: 13px;
              font-weight: 600;
              cursor: pointer;
              border: none;
              color: #fff;
              transition: opacity 0.2s ease;
            }

            .status_btn:hover {
              opacity: 0.9;
            }`
           }
        },
        'data/': {
          'opts.json': { raw: opts }
        }
      }
    }
  }
}
