// chat_view_header/chat_view_header.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { sdb, get } = statedb(fallback_module)

const chat_filter = require('chat_filter')

module.exports = chat_view_header

async function chat_view_header (opts = {}) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb
  
  const on = {
    style: inject,
    data: ondata,
    icons: iconject,

  }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  shadow.innerHTML = `
    <div class="chat_view_header"></div>
    <div class="chat_filter_panel"></div>
    <style></style>
  `
  const style = shadow.querySelector('style')
  const row = shadow.querySelector('.chat_view_header')
  const filter_panel = shadow.querySelector('.chat_filter_panel')
  
  
  let dricons = []
  const subs = await sdb.watch(onbatch)
  
  const chat_filter_component = await chat_filter(subs[0])
  filter_panel.appendChild(chat_filter_component)
  filter_panel.style.display = "none"

  return el

  function fail(data, type) {
    throw new Error('invalid message', { cause: { data, type } })
  }

  async function onbatch (batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(
        paths.map(path => drive.get(path).then(file => file.raw))
      )
      const func = on[type] || fail
      await func(data, type)
    }
  }

  function inject (data) {
    style.textContent = data[0]
  }

  async function ondata(data) {
  const { avatar, name, code } = data[0]
  row.innerHTML = `
    <div class="container-view-header">
      <div class="container-left">
        <div class="left-icon">
          ${dricons[0] || ''}
        </div>

        <div class="contact-heading">
          <div class="contact-avatar">
            <img src="${avatar}" alt="avatar" />
          </div>
          <div class="contact-info">
            <div class="contact-name">${name}</div>
            <div class="contact-code">${code}</div>
          </div>
        </div>
      </div>
      <div class="right-icon">
        ${dricons[1] || ''}
      </div>
    </div>
  `

  const right_icon = row.querySelector('.right-icon')
  right_icon.addEventListener('click', () => {
    const visible = filter_panel.style.display === "block"
    filter_panel.style.display = visible ? "none" : "block"
  })
}

   function iconject (data) {
    dricons = data
  }
}


function fallback_module () {

  return {
    api: fallback_instance,
    _: {
      'chat_filter': { $: '' },
    } 
  }
  function fallback_instance (opts) {
    if (!opts) opts = {}
    if (!opts.value) opts.value = {}
    
    const chat_filter = {
      mapping: {
        style: 'style',
        data: 'data',
        icons: 'icons'
      },
      0: {      }
    }
    return {
        drive: {
          'icons/': {
            'left-arrow.svg': {
              '$ref': 'left-arrow.svg'
            },
            'three-dot.svg':{
              '$ref': 'three-dot.svg'
            }
          },
          'style/': {
          'chat_view_header.css': { '$ref': 'chat_view_header.css' }
          },
          'data/': {
            'opts.json': {
              raw: opts || {}
            }
          }
        },
        _: { chat_filter }
    }
  }
}
