// switch_request/switch_request.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { sdb, get } = statedb(fallback_module)

const request_btc = require('request_btc')
const request_light = require('request_light')

module.exports = switch_request

async function switch_request(opts = {}, protocol) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = {
    style: inject,
    data: ondata,
    icons: iconject,
  }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  let dricons = []
  let showSendBtc = false
  let request_btc_comp = null
  let req_light_comp = null

  shadow.innerHTML = `
    <div class="parent-wrapper">
      <div class="pop-up"></div>

      <div class="switch-account-container">
        <div class="account-content"></div>
      </div>
    </div>

    <style></style>
  `

  const popup = shadow.querySelector('.pop-up')
  const container = shadow.querySelector('.switch-account-container')
  const content = shadow.querySelector('.account-content')
  const style = shadow.querySelector('style')

  popup.style.display = "none"

  // ONLY ONE SUBS EVER
  const subs = await sdb.watch(onbatch)

  // Build send_btc ONE time
  request_btc_comp = await request_btc(subs[0])
  request_btc_comp.classList.add("request-btc")

  req_light_comp = await request_light(subs[1])
  req_light_comp.classList.add("create-invoice")

  return el

  async function onbatch(batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(p => drive.get(p).then(f => f.raw)))
      const fn = on[type] || fail
      await fn(data, type)
    }
  }

  function inject(data) {
    style.textContent = data[0]
  }

  async function ondata(data) {
    const { btc, lightning } = data[0] || {}

    // If popup is open (send_btc visible), do NOT overwrite BTC/Lightning UI
    if (showSendBtc) return

    content.innerHTML = `
      <div class="account-container btc-container">
        <div class="btc-icon">${dricons[1] || '₿'} BTC</div>
      </div>

      <div class="account-container lightning-container">
        <div class="lightning-icon">${dricons[2] || '⚡'} Lightning</div>
      </div>
    `

    // BTC CLICK → toggle popup visibility
    content.querySelector('.btc-container').onclick = () => {
      toggle_btc_request()
    }

    content.querySelector('.lightning-container').onclick = () => {
      toggle_light_request()
    }
  }

  // Toggle open/close of send_btc
  function toggle_btc_request() {
    showSendBtc = !showSendBtc
    popup.classList.remove('hidden')

    if (showSendBtc) {
      popup.innerHTML = ""
      popup.appendChild(request_btc_comp)
      popup.style.display = "flex"

    } else {
      popup.style.display = "none"
      popup.innerHTML = ""
    }
  }

   // Toggle open/close of send_btc
  function toggle_light_request() {
    showSendBtc = !showSendBtc
    popup.classList.remove('hidden')

    if (showSendBtc) {
      popup.innerHTML = ""
      popup.appendChild(req_light_comp)
      popup.style.display = "flex"

    } else {
      popup.style.display = "none"
      popup.innerHTML = ""
    }
  }

  function iconject(data) {
    dricons = data
  }

  function fail(data, type) {
    throw new Error("invalid message", { cause: { data, type } })
  }
}



// ================== fallback ==================
function fallback_module() {
  return {
    api: fallback_instance,
    _: {
      'request_btc': { $: '' },
      'request_light': { $: '' }
    }
  }

  function fallback_instance(opts) {
    const request_btc = {
      mapping: { style: 'style', data: 'data', icons: 'icons' },
      0: {}
    }

    const request_light = {
      mapping: { style: 'style', data: 'data', icons: 'icons' },
      1: {}
    }

    return {
      drive: {
        'icons/': {
          'x.svg': { '$ref': 'x.svg' },
          'btc.svg': { '$ref': 'btc.svg' },
          'lightning.svg': { '$ref': 'lightning.svg' }
        },
        'style/': {
          'style.css': { '$ref': 'switch_request.css' }
        },
        'data/': {
          'opts.json': { raw: opts }
        }
      },
      _: { request_btc, request_light}
    }
  }
}
