// switch_request/switch_request.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { sdb, get } = statedb(fallback_module)

const request_btc = require('request_btc')
const request_light = require('request_light')

module.exports = switch_request

async function switch_request(opts = {}, protocol) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = {
    style: inject,
    data: ondata,
    icons: iconject,
  }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  let dricons = []
  let showSendBtc = false

  shadow.innerHTML = `
    <div class="parent-wrapper">
      <div class="switch-account-container">
        <div class="account-content"></div>
      </div>
    </div>
    <div class="request-btc hidden"></div>
    <div class="request-light hidden"></div>

    <style></style>
  `

  const content = shadow.querySelector('.account-content')
  const style = shadow.querySelector('style')
  const req_btc_el = shadow.querySelector('.request-btc')
  const req_light_el = shadow.querySelector('.request-light')
  
  const subs = await sdb.watch(onbatch)

  let btc_mounted = false
  let light_mounted = false



  return el

  async function onbatch(batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(p => drive.get(p).then(f => f.raw)))
      const fn = on[type] || fail
      await fn(data, type)
    }
  }

  function inject(data) {
    style.textContent = data[0]
  }

  async function ondata(data) {
    const { btc, lightning } = data[0] || {}

    if (showSendBtc) return

    content.innerHTML = `
      <div class="account-container btc-container">
        <div class="btc-icon">${dricons[1] || '₿'} BTC</div>
      </div>

      <div class="account-container lightning-container">
        <div class="lightning-icon">${dricons[2] || '⚡'} Lightning</div>
      </div>
    `
    const btc_container = content.querySelector('.btc-container')
    const lightning_container = content.querySelector('.lightning-container')

    btc_container.addEventListener('click', async () => {
      req_btc_el.classList.remove('hidden')

      if (!btc_mounted) {
        const history = await request_btc(subs[0], {
          onClose: () => req_btc_el.classList.add('hidden')
        })
        req_btc_el.append(history)
        btc_mounted = true
      }
    })
  
    lightning_container.addEventListener('click', async () => {
      req_light_el.classList.remove('hidden')

      if (!light_mounted) {
        const history = await request_light(subs[1], {
          onClose: () => req_light_el.classList.add('hidden')
        })
        req_light_el.append(history)
        light_mounted = true
      }
    })
  }



  function iconject(data) {
    dricons = data
  }

  function fail(data, type) {
    throw new Error("invalid message", { cause: { data, type } })
  }
}



// ================== fallback ==================
function fallback_module() {
  return {
    api: fallback_instance,
    _: {
      'request_btc': { $: '' },
      'request_light': { $: '' }
    }
  }

  function fallback_instance(opts) {
    if (!opts) opts = {}
    if (!opts.value) opts.value = {}

    const request_btc = {
      mapping: { style: 'style', data: 'data', icons: 'icons' },
      0: {}
    }

    const request_light = {
      mapping: { style: 'style', data: 'data', icons: 'icons' },
      1: {}
    }

    return {
      drive: {
        'icons/': {
          'x.svg': { '$ref': 'x.svg' },
          'btc.svg': { '$ref': 'btc.svg' },
          'lightning.svg': { '$ref': 'lightning.svg' }
        },
        'style/': {
          'style.css': { '$ref': 'switch_request.css' }
        },
        'data/': {
          'opts.json': { raw: opts || {} }
        }
      },
      _: { request_btc, request_light}
    }
  }
}
