// switch_request/switch_request.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { sdb, get } = statedb(fallback_module)

const request_btc = require('request_btc')
const request_light = require('request_light')

module.exports = switch_request

async function switch_request(opts = {}, protocol) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = {
    style: inject,
    data: ondata,
    icons: iconject,
  }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  let dricons = []
  let showSendBtc = false

  shadow.innerHTML = `
    <div class="parent-wrapper">
      <div class="pop-up-btc hidden"></div>
      <div class="pop-up-light hidden"></div>
      <div class="switch-account-container">
        <div class="account-content"></div>
      </div>
    </div>

    <style></style>
  `

  const popup = shadow.querySelector('.pop-up')
  const content = shadow.querySelector('.account-content')
  const style = shadow.querySelector('style')

  const subs = await sdb.watch(onbatch)

  const req_btc_comp = await request_btc(subs[0])
  const req_light_comp = await request_light(subs[1])

  shadow.querySelector('.pop-up-btc').prepend(req_btc_comp)
  shadow.querySelector('.pop-up-light').prepend(req_light_comp)

  return el

  async function onbatch(batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(p => drive.get(p).then(f => f.raw)))
      const fn = on[type] || fail
      await fn(data, type)
    }
  }

  function inject(data) {
    style.textContent = data[0]
  }

  async function ondata(data) {
    const { btc, lightning } = data[0] || {}

    if (showSendBtc) return

    content.innerHTML = `
      <div class="account-container btc-container">
        <div class="btc-icon">${dricons[1] || '₿'} BTC</div>
      </div>

      <div class="account-container lightning-container">
        <div class="lightning-icon">${dricons[2] || '⚡'} Lightning</div>
      </div>
    `

    content.querySelector('.btc-container').onclick = toggle_btc_request
    content.querySelector('.lightning-container').onclick = toggle_light_request
  }


  function openPopup(component) {
    showSendBtc = true
    popup.classList.remove("hidden")
    popup.style.display = "flex"

    popup.innerHTML = ""
    popup.appendChild(component)
  }

  function closePopup() {
    showSendBtc = false
    popup.classList.add("hidden")
    popup.style.display = "none"
    popup.innerHTML = ""
  }

  // Called by your BTC/Lightning icons
  function toggle_btc_request() {
    if (showSendBtc) return closePopup()
    openPopup(req_btc_comp)
  }

  function toggle_light_request() {
    if (showSendBtc) return closePopup()
    openPopup(req_light_comp)
  }

  function iconject(data) {
    dricons = data
  }

  function fail(data, type) {
    throw new Error("invalid message", { cause: { data, type } })
  }
}



// ================== fallback ==================
function fallback_module() {
  return {
    api: fallback_instance,
    _: {
      'request_btc': { $: '' },
      'request_light': { $: '' }
    }
  }

  function fallback_instance(opts) {
    const request_btc = {
      mapping: { style: 'style', data: 'data', icons: 'icons' },
      0: {}
    }

    const request_light = {
      mapping: { style: 'style', data: 'data', icons: 'icons' },
      1: {}
    }

    return {
      drive: {
        'icons/': {
          'x.svg': { '$ref': 'x.svg' },
          'btc.svg': { '$ref': 'btc.svg' },
          'lightning.svg': { '$ref': 'lightning.svg' }
        },
        'style/': {
          'style.css': { '$ref': 'switch_request.css' }
        },
        'data/': {
          'opts.json': { raw: opts }
        }
      },
      _: { request_btc, request_light}
    }
  }
}
