// share_invoice_via/share_invoice_via.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { sdb, get } = statedb(fallback_module)

module.exports = create_invoice_via

async function create_invoice_via (opts = {}) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = {
    style: inject_style,
    data: ondata,
    icons: iconject,
  }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  shadow.innerHTML = `
    <div class="invoice-container">
      <div class="header">Share invoice via</div>
      <div class="divider"></div>
      <div class="apps-row"></div>
    </div>
    <style></style>
  `

  const style = shadow.querySelector('style')
  const apps_row = shadow.querySelector('.apps-row')

  let dricons = []
  await sdb.watch(onbatch)

  return el

  function fail (data, type) {
    throw new Error('invalid message', { cause: { data, type } })
  }

  async function onbatch (batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(
        paths.map(path => drive.get(path).then(file => file.raw))
      )
      const func = on[type] || fail
      await func(data, type)
    }
  }

  function inject_style (data) {
    style.textContent = data[0]
  }

  async function ondata () {
    const app_names = ['Instagram', 'WhatsApp', 'Facebook', 'Binance', 'Meta', 'Trust']

    apps_row.innerHTML = app_names.map((name, i) => `
      <div class="app-item">
        <div class="icon-box">
          ${dricons[i] || ''}
        </div>
        <div class="app-name">${name}</div>
      </div>
    `).join('')
  }

  function iconject (data) {
    dricons = data
  }
}

function fallback_module () {
  return {
    api: fallback_instance,
  }
  function fallback_instance (opts) {
    return {
      drive: {
        'icons/': {
          'instagram.svg': { '$ref': 'instagram.svg' },
          'whatsapp.svg': { '$ref': 'whatsapp.svg' },
          'facebook.svg': { '$ref': 'facebook.svg' },
          'binance.svg': { '$ref': 'binance.svg' },
          'meta.svg': { '$ref': 'meta.svg' },
          'trustwallet.svg': { '$ref': 'trustwallet.svg' },
        },
        'style/': {
          'share_invoice_via.css': {
            '$ref': 'share_invoice_via.css'
          }
        },
        'data/': {
          'opts.json': { raw: opts }
        }
      }
    }
  }
}
