// switch_account/switch_account.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { sdb, get } = statedb(fallback_module)

// require pages early

module.exports = switch_account

async function switch_account (opts = {}, protocol) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = {
    style: inject,
    data: ondata,
    icons: iconject,
  }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  let dricons = []

  shadow.innerHTML = `
    <div class="switch-account-container">
      <div class="account-content"></div>
    </div>
    <style></style>
  `
  const style = shadow.querySelector('style')
  const content = shadow.querySelector('.account-content')

  const subs = await sdb.watch(onbatch)

  // protocol pattern:
  // protocol expects a function (sendToSwitch) and returns a function (sendFromSwitch).
  // child calls protocol(sendToSwitch) and receives send() to send messages up.
  const send = protocol ? protocol(function (messageToSwitch) {
    // parent -> switch_account messages (not used currently)
    console.log('[switch_account] message from parent ->', messageToSwitch)
  }) : null

  return el

  async function onbatch (batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(
        paths.map(path => drive.get(path).then(file => file.raw))
      )
      const func = on[type] || fail
      await func(data, type)
    }
  }

  function inject (data) {
    style.textContent = data[0]
  }

  async function ondata (data) {
    const { btc, lightning } = data[0] || {}

    content.innerHTML = `
      <div class="container-title">
        <div class="title">Switch Account</div>
        <div class="close-icon">${dricons[0] || '✕'}</div>   
      </div>
      <div class="account-container btc-container">
        <div class="btc-icon">${dricons[1] || '₿'} BTC</div>
        <div class="btc-amount">${typeof btc !== 'undefined' ? parseFloat(btc).toFixed(4) : '0.0000'}</div>       
      </div>
      <div class="account-container lightning-container">
        <div class="lightning-icon">${dricons[2] || '⚡'} Lightning</div>
        <div class="lightning-amount">${typeof lightning !== 'undefined' ? parseFloat(lightning).toFixed(4) : '0.0000'}</div>       
      </div>
    `

    // Close button
    const close_btn = content.querySelector('.close-icon')
    if (close_btn) {
      close_btn.onclick = () => {
        const dropdown = el.parentNode
        if (dropdown) dropdown.classList.add('hidden')
      }
    }

    // BTC Click: send protocol message up
    const btc_container = content.querySelector('.btc-container')
    if (btc_container) {
      btc_container.onclick = () => {
        if (send) send({ type: 'switch', data: 'btc' })
      }
    }

    // Lightning Click: send protocol message up
    const lightning_container = content.querySelector('.lightning-container')
    if (lightning_container) {
      lightning_container.onclick = () => {
        if (send) send({ type: 'switch', data: 'lightning' })
      }
    }
  }

  function fail (data, type) {
    throw new Error('invalid message', { cause: { data, type } })
  }

  function iconject (data) {
    dricons = data
  }
}

// ================== fallback ==================
function fallback_module () {
  return {
    api: fallback_instance,
  }

  function fallback_instance (opts) {
    if (!opts) opts = {}
    if (!opts.value) opts.value = {}

    return {
      drive: {
        'icons/': {
          'x.svg': { '$ref': 'x.svg' },
          'btc.svg': { '$ref': 'btc.svg' },
          'lightning.svg': { '$ref': 'lightning.svg' }
        },
        'style/': {
          'style.css': { '$ref': 'switch_account.css' }
        },
        'data/': {
          'opts.json': { raw: opts || {} }
        }
      },
    }
  }
}
