// chat_view/chat_view.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { sdb, get } = statedb(fallback_module)

const button = require('button')
const btc_req_msg = require('btc_req_msg')
const switch_request = require('switch_request')
const switch_send = require('switch_send')
const chat_filter = require('chat_filter')

module.exports = chat_view

async function chat_view(opts = {}, { onClose } = {}) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = {
    style: inject,
    data: ondata,
    icons: iconject,
  }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  let dricons = []

  shadow.innerHTML = `
    <div class="chat-view-container">
      <div class="chat-view">
        <div class="chat-view-header"></div>
        <div class="chat_filter_panel"></div>
        <div class="chat-messages"></div>
      </div>
      <div class="buttons">
        <div class="request">       
          <div class="popup-box-request"></div>
          <div class="request-button"></div>
        </div>
        <div class="send">       
          <div class="popup-box-send"></div>
          <div class="send-button"></div>
        </div>
      </div>
    </div>
    <style></style>
  `

  const style = shadow.querySelector('style')

  const chat_view_container = shadow.querySelector('.chat-view-container')
  const header_component = shadow.querySelector('.chat-view-header')
  const send_button_component = shadow.querySelector('.send-button')
  const request_button_component = shadow.querySelector('.request-button')
  const messages_component = shadow.querySelector('.chat-messages')
  const buttons_container = shadow.querySelector('.buttons')
  const filter_panel = shadow.querySelector('.chat_filter_panel')
  const popup_box_request = shadow.querySelector('.popup-box-request')
  const popup_box_send = shadow.querySelector('.popup-box-send')
  
  const subs = await sdb.watch(onbatch)

  const chat_filter_component = await chat_filter(subs[0])
  const send_button = await button(subs[1])
  const request_button = await button(subs[2])
  const switch_request_component = await switch_request(subs[3])
  const switch_send_component = await switch_send(subs[4])
  const btc_req_msg1_component = await btc_req_msg(subs[5])
  const btc_req_msg2_component = await btc_req_msg(subs[6])
  const btc_req_msg3_component = await btc_req_msg(subs[7])

  send_button_component.replaceWith(send_button)
  request_button_component.replaceWith(request_button)
  messages_component.appendChild(btc_req_msg1_component)
  messages_component.appendChild(btc_req_msg2_component)
  messages_component.appendChild(btc_req_msg3_component)
  filter_panel.appendChild(chat_filter_component)
  filter_panel.style.display = "none"


  popup_box_send.appendChild(switch_send_component)
  popup_box_request.appendChild(switch_request_component)


  let openPopup = null

  function togglePopup(target) {
    if (openPopup === target) {
      target.style.display = 'none'
      openPopup = null
      return
    }

    if (openPopup) openPopup.style.display = 'none'

    target.style.display = 'block'
    openPopup = target
  }

  send_button.addEventListener('click', () => togglePopup(popup_box_send))
  request_button.addEventListener('click', () => togglePopup(popup_box_request))

  shadow.addEventListener('click', (e) => {
    if (!popup_box_send.contains(e.target) &&
        !popup_box_request.contains(e.target) &&
        e.target !== send_button &&
        e.target !== request_button) 
    {
      if (openPopup) openPopup.style.display = 'none'
      openPopup = null
    }
  })

  return el

  function fail(data, type) {
    throw new Error('invalid message', { cause: { data, type } })
  }

  async function onbatch(batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(
        paths.map(path => drive.get(path).then(f => f.raw))
      )
      const func = on[type] || fail
      await func(data, type)
    }
  }

  function inject(data) {
    style.textContent = data[0]
  }

  async function ondata(data) {
    const { avatar, name, code } = data[0] || {}
    
    header_component.innerHTML = `
      <div class="container-view-header">
        <div class="container-left">
          <div class="left-icon">
            ${dricons[0] || ''}
          </div>

          <div class="contact-heading">
            <div class="contact-avatar">
              <img src="${avatar}" alt="avatar" />
            </div>
            <div class="contact-info">
              <div class="contact-name">${name}</div>
              <div class="contact-code">${code}</div>
            </div>
          </div>
        </div>
        <div class="right-icon">
          ${dricons[1] || ''}
        </div>
      </div>
    `

    const right_icon = header_component.querySelector('.right-icon')
    right_icon.addEventListener('click', () => {
      const visible = filter_panel.style.display === "block"
      filter_panel.style.display = visible ? "none" : "block"
    })

    const left_icon = header_component.querySelector('.left-icon')
    
    // hide overlay on close
    left_icon.addEventListener('click', () => {
      if (onClose) onClose()
    })
  }

  function iconject (data) {
    dricons = data
  }
}

function fallback_module() {
  return {
    api,
    _: {
      'chat_filter': { $: '' },
      'button': { $: '' },
      'btc_req_msg': { $: '' },
      'switch_request': { $: '' },
      'switch_send': { $: '' }
    }
  }

  function api(opts) {
    if (!opts) opts = {}
    if (!opts.value) opts.value = {}

    const chat_filter = {
      mapping: { style: 'style', data: 'data', icons: 'icons' },
      0: {}
    }

    const button = {
      mapping: { style: 'style', data: 'data' },
      1: { label: 'Send' },
      2: { label: 'Request' }
    }

    const switch_request = {
      mapping: { style: 'style', data: 'data', icons: 'icons' },
      3: opts.value
    }

    const switch_send = {
      mapping: { style: 'style', data: 'data', icons: 'icons' },
      4: opts.value
    }

    // BTC messages with separate statuses
    const btc_req_msg = {
      mapping: { style: 'style', data: 'data', icons: 'icons' },
      5: { ...opts.value, 
        date: "25 Dec 2026",
        avatar: "https://tse4.mm.bing.net/th/id/OIP.bdn3Kne-OZLwGM8Uoq5-7gHaHa?w=512&h=512&rs=1&pid=ImgDetMain&o=7&rm=3",
        name: 'David Clark',
        amount: 0.0054,
        status: "send", 
        is_me: false
      },
      6: { ...opts.value,    
        date: "25 Sep 2025",
        avatar: "https://tse4.mm.bing.net/th/id/OIP.bdn3Kne-OZLwGM8Uoq5-7gHaHa?w=512&h=512&rs=1&pid=ImgDetMain&o=7&rm=3",
        name: 'David Clark',
        amount: 0.0052,
        status: "paid", 
        is_me: false},
      7: { ...opts.value, 
        date: "25 Dec 2027",
        avatar: "https://tse4.mm.bing.net/th/id/OIP.bdn3Kne-OZLwGM8Uoq5-7gHaHa?w=512&h=512&rs=1&pid=ImgDetMain&o=7&rm=3",
        name: 'David Clark',
        amount: 0.0054,
        status: "paid", 
        is_me: false
       }
    }

    return {
      drive: {
        'icons/': {
          'left-arrow.svg': {
            '$ref': 'left-arrow.svg'
          },
          'three-dot.svg':{
            '$ref': 'three-dot.svg'
          }
        },
        'style/': {
          'chat_view.css': { '$ref': 'chat_view.css' }
        },
        'data/': {
          'opts.json': { raw: opts.value || {} }
        }
      },
      _: {
        chat_filter,
        button,
        switch_request,
        switch_send,
        btc_req_msg
      }
    }
  }
}
