// chat_view/chat_view.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { sdb, get } = statedb(fallback_module)

const chat_view_header = require('chat_view_header')
const button = require('button')
const btc_req_msg = require('btc_req_msg')
const switch_request = require('switch_request')
const switch_send = require('switch_send')

module.exports = chat_view

async function chat_view(opts = {}) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = {
    style: inject,
    data: ondata
  }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  shadow.innerHTML = `
    <div class="component-label">Chat View</div>
    <div class="chat-view-container">
      <div class="chat-view">
        <div class="chat-view-header"></div>
        <div class="chat-messages"></div>
      </div>
      <div class="buttons">
        <div class="request-button"></div>
        <div class="send-button"></div>
      </div>
    </div>
    <style></style>
  `

  const style = shadow.querySelector('style')

  const header_component = shadow.querySelector('.chat-view-header')
  const send_button_component = shadow.querySelector('.send-button')
  const request_button_component = shadow.querySelector('.request-button')
  const messages_component = shadow.querySelector('.chat-messages')
  const buttons_container = shadow.querySelector('.buttons')

  
  const subs = await sdb.watch(onbatch)

  const header = await chat_view_header(subs[0])
  const send_button = await button(subs[1])
  const request_button = await button(subs[2])
  const switch_request_component = await switch_request(subs[3])
  const switch_send_component = await switch_send(subs[4])
  const btc_req_msg1_component = await btc_req_msg(subs[5])
  const btc_req_msg2_component = await btc_req_msg(subs[6])
  const btc_req_msg3_component = await btc_req_msg(subs[7])

  send_button_component.replaceWith(send_button)
  request_button_component.replaceWith(request_button)
  header_component.append(header)
  messages_component.appendChild(btc_req_msg1_component)
  messages_component.appendChild(btc_req_msg2_component)
  messages_component.appendChild(btc_req_msg3_component)

  // POPUPS ABOVE BUTTONS
  const send_popup = document.createElement('div')
  send_popup.className = 'popup-box-send'
  send_popup.append(switch_send_component)

  const request_popup = document.createElement('div')
  request_popup.className = 'popup-box-request'
  request_popup.appendChild(switch_request_component)

  buttons_container.parentNode.insertBefore(send_popup, buttons_container)
  buttons_container.parentNode.insertBefore(request_popup, buttons_container) 

  let openPopup = null

  function togglePopup(target) {
    if (openPopup === target) {
      target.style.display = 'none'
      openPopup = null
      return
    }

    if (openPopup) openPopup.style.display = 'none'

    target.style.display = 'block'
    openPopup = target
  }

  send_button.addEventListener('click', () => togglePopup(send_popup))
  request_button.addEventListener('click', () => togglePopup(request_popup))

  shadow.addEventListener('click', (e) => {
    if (!send_popup.contains(e.target) &&
        !request_popup.contains(e.target) &&
        e.target !== send_button &&
        e.target !== request_button) 
    {
      if (openPopup) openPopup.style.display = 'none'
      openPopup = null
    }
  })

  return el

  function fail(data, type) {
    throw new Error('invalid message', { cause: { data, type } })
  }

  async function onbatch(batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(
        paths.map(path => drive.get(path).then(file => file.raw))
      )
      const func = on[type] || fail
      await func(data, type)
    }
  }

  function inject(data) {
    style.textContent = data[0]
  }

  async function ondata(data) {}
}

function fallback_module() {
  return {
    api,
    _: {
      'chat_view_header': { $: '' },
      'button': { $: '' },
      'btc_req_msg': { $: '' },
      'switch_request': { $: '' },
      'switch_send': { $: '' }
    }
  }

  function api(opts) {

    const chat_view_header = {
      mapping: { style: 'style', data: 'data', icons: 'icons' },
      0: opts.value
    }

    const button = {
      mapping: { style: 'style', data: 'data' },
      1: { label: 'Send' },
      2: { label: 'Request' }
    }

    const switch_request = {
      mapping: { style: 'style', data: 'data', icons: 'icons' },
      3: opts.value
    }

    const switch_send = {
      mapping: { style: 'style', data: 'data', icons: 'icons' },
      4: opts.value
    }

    // BTC messages with separate statuses
    const btc_req_msg = {
      mapping: { style: 'style', data: 'data', icons: 'icons' },
      5: { ...opts.value, status: 'expired' },
      6: { ...opts.value, status: 'paid' },
      7: { ...opts.value, status: 'send' }
    }

    return {
      drive: {
        'style/': {
          'chat_view.css': { '$ref': 'chat_view.css' }
        },
        'data/': {
          'opts.json': { raw: opts }
        }
      },
      _: {
        chat_view_header,
        button,
        switch_request,
        switch_send,
        btc_req_msg
      }
    }
  }
}
