// send_btc/send_btc.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { sdb, get } = statedb(fallback_module)

const create_button = require('button')
const btc_input_card = require('btc_input_card')
const templates = require('templates')

module.exports = send_btc

async function send_btc(opts = {}) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = {
    style: inject,
    data: ondata,
    icons: iconject,
  }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  let dricons = []

  shadow.innerHTML = `
    <div class="create-invoice-container">
      <div class="create-invoice-header">  
        <div class="title-container"> 
          <div class="create-invoice-header">Send BTC</div>
          <div class="btc-icon"></div>
        </div>  
        <div class="x-icon"></div>
      </div>
      <div class="btc-input-card"></div>
      <div class="divider">
        <div class="sub-divider"></div>
        <div>Or</div>
        <div class="sub-divider"></div>
      </div>      
      <div class="templates-heading">Use Templates</div>
      <div class="template1"></div>
      <div class="template2"></div>
      <div class="template3"></div>
      <div class="create_button"></div>
    </div>
    <style></style>
  `

  const style = shadow.querySelector('style')
  const create_button_component = shadow.querySelector('.create_button')
  const btc_input_card_component = shadow.querySelector('.btc-input-card')
  const template1_container = shadow.querySelector('.template1')
  const template2_container = shadow.querySelector('.template2')
  const template3_container = shadow.querySelector('.template3')

  const subs = await sdb.watch(onbatch)

  const button_component = await create_button(subs[0])
  const btc_component = await btc_input_card(subs[1])
  const templates_component1 = await templates(subs[2])
  const templates_component2 = await templates(subs[3])
  const templates_component3 = await templates(subs[4])


  create_button_component.append(button_component)
  btc_input_card_component.append(btc_component)
  template1_container.append(templates_component1)
  template2_container.append(templates_component2)
  template3_container.append(templates_component3)

  const closeBtn = shadow.querySelector('.x-icon')
  if (closeBtn) {
    closeBtn.onclick = () => {
      const dropdown = el.parentNode
      if (dropdown) {
        dropdown.classList.add('hidden')   
      }
    }
  }

  return el

  function fail(data, type) {
    throw new Error('invalid message', { cause: { data, type } })
  }

  async function onbatch(batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(path => drive.get(path).then(file => file.raw)))
      const func = on[type] || fail
      await func(data, type)
    }
  }

  function inject(data) {
    style.textContent = data[0]
  }

  function iconject (data) {
    dricons = data

    const btc_icon = shadow.querySelector('.btc-icon')
    const x_icon = shadow.querySelector('.x-icon')

    btc_icon.innerHTML = dricons[0]
    x_icon.innerHTML = dricons[1]
  }

  async function ondata(data) {
    
  }
}

function fallback_module() {
  return {
    api,
    _: {
      'button': { $: '' },
      'btc_input_card': { $: '' },
      'templates': { $: '' },
    }
  }

  function api(opts) {
    
    const button = {
      mapping: {
        style: 'style',
        data: 'data',
      },
      0: {
        label: 'Send'
      }    
    }
    
    const btc_input_card = {
      mapping: {
        style: 'style',
        data: 'data',
      },
      1: {
        balance: 0.0024, 
      }  
    } 

    const templates = {
      mapping: {
        style: 'style',
        data: 'data',
        icons: 'icons'
      },
      2: {
        date:"2025-05-10",
        btc: 0.012,
        usd: 200
      },
      3: {
        date:"2024-12-25",
        btc: 0.005,
        usd: 80
      },
      4: {
        date:"2024-01-01",
        btc: 0.01,
        usd: 150
      }
    }

    return {
      drive: {
        'icons/': {
          'btc.svg': {
            '$ref': 'btc.svg'
          },
          'x.svg': {
            '$ref': 'x.svg'
          },
        },
        'style/': {
          'send_btc.css': {
            '$ref': 'send_btc.css'
          }
        },
        'data/': {
          'opts.json': {
            raw: opts
          }
        }
      },
      _: {
        button,
        btc_input_card,
        templates      
      }
    }
  }
}
