// chat_filter/chat_filter.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { sdb, get } = statedb(fallback_module)

module.exports = chat_filter

async function chat_filter(opts = {}) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = {
    style: inject,
    data: ondata,
    icons: iconject,
  }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  let dricons = [] // arrow, orange, green, red icons

  shadow.innerHTML = `
    <div class="cf-wrap">
      <div class="cf-item main-head" data-key="me">
        <span class="arrow-slot"></span>
        <span>Request by me</span>
      </div>

      <div class="section-me hidden">
        <div class="cf-item sub-head" data-key="me-btc">
          <span class="arrow-slot"></span>
          <span>Bitcoin</span>
        </div>
        <div class="sub-me-btc hidden cf-status"></div>

        <div class="cf-item sub-head" data-key="me-ln">
          <span class="arrow-slot"></span>
          <span>Lightning</span>
        </div>
        <div class="sub-me-ln hidden cf-status"></div>
      </div>

      <div class="cf-item main-head" data-key="luis">
        <span class="arrow-slot"></span>
        <span>Request by luis</span>
      </div>

      <div class="section-luis hidden">
        <div class="cf-item sub-head" data-key="luis-btc">
          <span class="arrow-slot"></span>
          <span>Bitcoin</span>
        </div>
        <div class="sub-luis-btc hidden cf-status"></div>

        <div class="cf-item sub-head" data-key="luis-ln">
          <span class="arrow-slot"></span>
          <span>Lightning</span>
        </div>
        <div class="sub-luis-ln hidden cf-status"></div>
      </div>
    </div>
    <style></style>
  `

  const style = shadow.querySelector('style')

  const sections = {
    me: shadow.querySelector('.section-me'),
    'me-btc': shadow.querySelector('.sub-me-btc'),
    'me-ln': shadow.querySelector('.sub-me-ln'),
    luis: shadow.querySelector('.section-luis'),
    'luis-btc': shadow.querySelector('.sub-luis-btc'),
    'luis-ln': shadow.querySelector('.sub-luis-ln')
  }

  const subs = await sdb.watch(onbatch)

  // click listeners
  shadow.querySelectorAll('.cf-item').forEach(item => {
    item.onclick = () => toggleSection(item.dataset.key)
  })

  function toggleSection(key) {
    const box = sections[key]
    if (!box) return

    box.classList.toggle('hidden')

    const arrowSlot = shadow.querySelector(`.cf-item[data-key="${key}"] .arrow-slot`)
    if (arrowSlot && dricons.length > 0) {
      arrowSlot.innerHTML = box.classList.contains('hidden') ? dricons[0] : dricons[0].replace('rotate(0)', 'rotate(90deg)')
    }
  }

  return el

  function fail(data, type) {
    throw new Error('invalid message', { cause: { data, type } })
  }

  async function onbatch(batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(path => drive.get(path).then(file => file.raw)))
      const func = on[type] || fail
      await func(data, type)
    }
  }

  function inject(data) {
    style.textContent = data[0]
  }

  function iconject(data) {
    dricons = data

    // insert icons into arrow slots
    shadow.querySelectorAll('.arrow-slot').forEach(slot => {
      slot.innerHTML = dricons[0] // arrow icon
    })

    // fill status sections
    fillStatuses('sub-me-btc', false)
    fillStatuses('sub-me-ln', true)
    fillStatuses('sub-luis-btc', false)
    fillStatuses('sub-luis-ln', true)
  }

  function fillStatuses(key, includeExpired) {
    const box = shadow.querySelector(`.${key}`)
    if (!box) return

    box.innerHTML = `
      <div><span class="dot">${dricons[1]}</span><span>Pending</span></div>
      <div><span class="dot">${dricons[2]}</span><span>Paid</span></div>
      ${includeExpired ? `<div><span class="dot">${dricons[3]}</span><span>Expired</span></div>` : ''}
    `
  }

  async function ondata(data) {}
}

function fallback_module () {
  return {
    api: fallback_instance,
  }

  function fallback_instance (opts) {
    if (!opts) opts = {}
    if (!opts.value) opts.value = {}
    
    return {
      drive: {
       'icons/': {
          'arrow_down.svg': { '$ref': 'arrow_down.svg' },
          'orange_dot.svg': { '$ref': 'orange_dot.svg' },
          'green_dot.svg': { '$ref': 'green_dot.svg' },
          'red_dot.svg': { '$ref': 'red_dot.svg' },
        },
        'style/': {
          'chat_filter.css': { '$ref': 'chat_filter.css' }
        },
        'data/': {
          'opts.json': { raw: opts || {} }
        }
      },
    }
  }
}
