// add_new_contact/add_new_contact.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { sdb, get } = statedb(fallback_module)

const create_button = require('button')
const input_field = require('input_field')

module.exports = add_new_contact

async function add_new_contact(opts = {}) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = {
    style: inject,
    data: ondata,
    icons: iconject,
  }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  let dricons = []

  shadow.innerHTML = `
    <div class="create-invoice-container">
      <div class="create-invoice-header">  
        <div class="title-container"> 
          <div class="create-invoice-header">Add new contact</div>
        </div>  
        <div class="x-icon"></div>
      </div>
  
      <div class="extra-input1"></div>
      <div class="extra-input2"></div>
      <div class="connect_button"></div>
    </div>
    <style></style>
  `

  const style = shadow.querySelector('style')

  const create_button_component = shadow.querySelector('.connect_button')
  const input1_component = shadow.querySelector('.extra-input1')
  const input2_component = shadow.querySelector('.extra-input2')


  const subs = await sdb.watch(onbatch)

  const button_component = await create_button(subs[0])
  const input1 = await input_field(subs[1])
  const input2 = await input_field(subs[2])

  create_button_component.append(button_component)
  input1_component.append(input1)
  input2_component.append(input2)

  const closeBtn = shadow.querySelector('.x-icon')
  if (closeBtn) {
    closeBtn.onclick = () => {
      const dropdown = el.parentNode
      if (dropdown) {
        dropdown.classList.add('hidden')   
      }
    }
  }

  return el

  function fail(data, type) {
    throw new Error('invalid message', { cause: { data, type } })
  }

  async function onbatch(batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(path => drive.get(path).then(file => file.raw)))
      const func = on[type] || fail
      await func(data, type)
    }
  }

  function inject(data) {
    style.textContent = data[0]
  }

  function iconject (data) {
    dricons = data

    const x_icon = shadow.querySelector('.x-icon')
    x_icon.innerHTML = dricons[1]
  }

  async function ondata(data) {
    
  }
}

function fallback_module() {
  return {
    api,
    _: {
      'button': { $: '' },
      'input_field': { $: '' },
    }
  }

  function api(opts) {
    
    const button = {
      mapping: {
        style: 'style',
        data: 'data',
      },
      0: {
        label: 'Connect'
      }    
    } 

    const input_field = {
      mapping: {
        style: 'style',
        data: 'data',
        icons: 'icons'
      },
      1: {
        header: 'Nickname',
        placeholder: 'Luis Fedrick',
      },
      2:{
        header: 'Paste the invite code',
        placeholder: '221341'
      }
    }

    return {
      drive: {
        'icons/': {
          'lightning.svg': {
            '$ref': 'lightning.svg'
          },
          'x.svg': {
            '$ref': 'x.svg'
          },
        },
        'style/': {
          'add_new_contact.css': {
            '$ref': 'add_new_contact.css'
          }
        },
        'data/': {
          'opts.json': {
            raw: opts
          }
        }
      },
      _: {
        button,
        input_field,
      }
    }
  }
}
