// more_menu/more_menu.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { sdb, get } = statedb(fallback_module)

const btc_nodes = require('btc_nodes')

module.exports = more_menu

async function more_menu(opts = {}) {
  const { sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = { style: inject, data: ondata, icons: iconject }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'open' })

  let icons = []

  shadow.innerHTML = `
    <div class="more-menu-container"></div>
    <div class="nodes-container hidden"></div>
    <style></style>
  `

  const style = shadow.querySelector('style')
  const menuContainer = shadow.querySelector('.more-menu-container')
  const nodesContainer = shadow.querySelector('.nodes-container')

  const subs = await sdb.watch(onbatch)

  // Render menu options
  ondata()

  let nodesMounted = false

  // Event delegation: click on Nodes
  menuContainer.addEventListener('click', async (e) => {
    const target = e.target.closest('.option-container')
    if (!target) return

    const label = target.querySelector('.option-label')?.textContent
    if (label === 'Nodes') {
      nodesContainer.classList.remove('hidden')

      if (!nodesMounted) {
        const nodesComponent = await btc_nodes(subs[0], {
          onClose: () => nodesContainer.classList.add('hidden')
        })
        nodesContainer.append(nodesComponent)
        nodesMounted = true
      }
    }
  })

  return el

  function fail(data, type) {
    throw new Error('invalid message', { cause: { data, type } })
  }

  async function onbatch(batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(path => drive.get(path).then(f => f.raw)))
      const func = on[type] || fail
      await func(data, type)
    }
  }

  function inject(data) {
    style.textContent = data[0]
  }

  function iconject(data) {
    icons = data
    ondata()
  }

  function ondata() {
    menuContainer.innerHTML = `
      <div class="container-title">
        <div class="title">More</div>
      </div>
      <div class="option-container">
        <div class="dot-icon">${icons[1] || ''}</div>
        <div class="option-label">Nodes</div>
      </div>
      <div class="option-container">
        <div class="dot-icon">${icons[1] || ''}</div>
        <div class="option-label">Peers</div>
      </div>
    `
  }
}

function fallback_module() {

  return {
    api: fallback_instance,
    _: { btc_nodes: { $: '' } }
  }

  function fallback_instance(opts) {
    if (!opts) opts = {}
    if (!opts.value) opts.value = {}
 
    const btc_nodes = {
      mapping: { icons: 'icons', style: 'style', data: 'data' },
      0: {}
    }

    return {
      drive: {
        'icons/': {
          'x.svg': { '$ref': 'x.svg' },
          'dot.svg': { '$ref': 'dot.svg' }
        },
        'style/': { 'style.css': { '$ref': 'more_menu.css' } },
        'data/': { 'opts.json': { raw: opts || {} } }
      },
      _: { btc_nodes }
    }
  }
}
