// pay_invoice_confirmation/pay_invoice_confirmation.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { sdb, get } = statedb(fallback_module)

const receipt_row = require('receipt_row')
const send_button = require('button') 

module.exports = pay_invoice_confirmation

async function pay_invoice_confirmation(opts = {}, { onClose } = {}) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = {
    style: inject,
    data: ondata,
    icons: iconject,
  }

  let dricons = []

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  shadow.innerHTML = `
    <div class="invoice-card">
      <div class="invoice-header">
        <div class="invoice-title-section">
          <div class="invoice-confirmation">Confirmation</div>
          <div class="invoice-amount">
            <div class="light-icon-small"></div>
            <div class="invoice-value" id="invoice-value">0.0000</div>
          </div>       
        </div>
        <div class="close-icon"></div>
      </div>
      <div class="receipt-rows"></div>
      <div class="send-button"></div>
      <style></style>
    </div>
    `

  const style = shadow.querySelector('style')
  const rows_el = shadow.querySelector('.receipt-rows')
  const send_button_container = shadow.querySelector('.send-button')
  const close_icon = shadow.querySelector('.close-icon')

  const subs = await sdb.watch(onbatch)


  if (subs.length > 0) {
    const button = await send_button(subs[0])
    send_button_container.replaceWith(button)
  }

  for (let i = 1; i < subs.length; i++) {
    const row = await receipt_row(subs[i]) 
    rows_el.append(row)
  }


  // close on click
  close_icon.addEventListener('click', () => {
    if (onClose) onClose()
  })

  return el

  function fail(data, type) {
    throw new Error('invalid message', { cause: { data, type } })
  }

  async function onbatch(batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(path => drive.get(path).then(file => file.raw)))
      const func = on[type] || fail
      await func(data, type)
    }
  }


  function inject(data) {
    style.textContent = data[0]
  }

  function iconject(data) {
    dricons = data
    const light_icon = shadow.querySelectorAll('.light-icon-small')
    const close_icon = shadow.querySelector('.close-icon')

    light_icon.forEach(el => el.innerHTML = dricons[0])
    close_icon.innerHTML = dricons[1]                  
  }

  async function ondata(data) {

    const {value} = data?.[0] || {}
    const btc_value = value[0].value
    const valueEl = shadow.querySelector('.invoice-value')
    
    valueEl.textContent = btc_value

  }
}


function fallback_module() {
  return {
    api,
    _: {
      'button': { $: '' },
      'receipt_row': { 
        $: '' 
      }
    }
  }

  function api(opts) {
    if (!opts) opts = {}
    if (!opts.value) opts.value = {}
 
    const button = {
      mapping: {
        style: 'style',
        data: 'data',
        icons: 'icons'
      },
      0: {
        label: 'Pay invoice'
      }
    }
    const receipt_row = {
      mapping: {
        style: 'style',
        data: 'data',
        icons: 'icons'
      }
    }
    opts.value.forEach((row, index) => {
      receipt_row[index + 1] = row
    })

    return {
      drive: {
        'icons/': {
          'lightning.svg': { '$ref': 'lightning.svg' },
          'x.svg': { '$ref': 'x.svg' }
        },
        'style/': {
          'pay_invoice_confirmation.css': { '$ref': 'pay_invoice_confirmation.css' }
        },
        'data/': {
          'opts.json': { raw: opts || {} }
        }
      },
      _: { button, receipt_row }
    }
  }
}
