// action_buttons/action_buttons.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { sdb, get } = statedb(fallback_module)

const general_button = require('general_button')
const send_btc = require('send_btc')
const receive_btc = require('receive_btc')

module.exports = action_buttons

async function action_buttons(opts = {}, protocol) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = { style: inject, data: ondata }
  const _ = { send_general: null, receive_general: null }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  shadow.innerHTML = `
    <div class="action-buttons-container">
      <div class="send-receive-buttons">
        <div id="send-button-container"></div>
        <div id="receive-button-container"></div>
      </div>
      <div class="overlay send-overlay hidden"></div>
      <div class="overlay receive-overlay hidden"></div>
    </div>
    <style></style>
  `

  const style = shadow.querySelector('style')
  const send_container = shadow.querySelector('#send-button-container')
  const receive_container = shadow.querySelector('#receive-button-container')
  const send_overlay = shadow.querySelector('.send-overlay')
  const receive_overlay = shadow.querySelector('.receive-overlay')

  const subs = await sdb.watch(onbatch)

  const sendUp = protocol ? protocol(msg => console.log('[action_buttons]:', msg)) : null

  // -------------------- Buttons --------------------
  const send_button = await general_button(subs[0], button_protocol('send_general'))
  const receive_button = await general_button(subs[1], button_protocol('receive_general'))

  send_container.appendChild(send_button)
  receive_container.appendChild(receive_button)

  send_button._action = 'send_message'
  receive_button._action = 'receive_message'

  let send_el = null, receive_el = null

  // -------------------- Overlay helpers --------------------
  function close_all_overlays(except) {
    if (except !== 'send') send_overlay.classList.add('hidden')
    if (except !== 'receive') receive_overlay.classList.add('hidden')
  }

  function handle_outside_click(event) {
    const target = event.target
    const clicked_inside =
      send_overlay.contains(target) || receive_overlay.contains(target)
    const clicked_button =
      send_button.contains(target) || receive_button.contains(target)

    if (!clicked_inside && !clicked_button) close_all_overlays(null)
  }

  shadow.addEventListener('click', handle_outside_click)

  // -------------------- Button Clicks --------------------
  send_button.onclick = async (e) => {
    e.stopPropagation()
    if (send_button._action !== 'send_message') return
    close_all_overlays('send')

    if (!send_el) {
      send_el = await send_btc(subs[2], { onClose: () => send_overlay.classList.add('hidden') })
      send_overlay.appendChild(send_el)
    }

    send_overlay.classList.remove('hidden')
    sendUp?.({ from: 'action_buttons', type: 'send', data: null })
  }

  receive_button.onclick = async (e) => {
    e.stopPropagation()
    if (receive_button._action !== 'receive_message') return
    close_all_overlays('receive')

    if (!receive_el) {
      receive_el = await receive_btc(subs[3], { onClose: () => receive_overlay.classList.add('hidden') })
      receive_overlay.appendChild(receive_el)
    }

    receive_overlay.classList.remove('hidden')
    sendUp?.({ from: 'action_buttons', type: 'receive', data: null })
  }

  // -------------------- Initial Config --------------------
  _.send_general?.({ type: 'button_name', data: { name: 'Send', action: 'send_message' } })
  _.receive_general?.({ type: 'button_name', data: { name: 'Receive', action: 'receive_message' } })

  return el

  // -------------------- Helpers --------------------
  function fail(data, type) { throw new Error('Invalid message type', { cause: { data, type } }) }
  async function onbatch(batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(p => drive.get(p).then(f => f.raw)))
      const handler = on[type] || fail
      handler(data, type)
    }
  }
  function inject(data) { style.textContent = data[0] }
  async function ondata(data) { const buttonData = data[0]?.value || {} }
  function button_protocol(key) { return send => { _[key] = send; return send } }
}

// -------------------- Fallback for STATE --------------------
function fallback_module() {
  
  return {
    api,
    _: { 'general_button': { $: '' }, 'send_btc': { $: '' }, 'receive_btc': { $: '' } }
  }

  function api(opts) {
    if (!opts) opts = {}
    if (!opts.value) opts.value = {}

    const general_button = { mapping: { style: 'style', data: 'data' }, 0: {}, 1: {} }
    const send_btc = { mapping: { style: 'style', data: 'data', icons: 'icons' }, 2: '' }
    const receive_btc = { mapping: { style: 'style', data: 'data', icons: 'icons' }, 3: '' }

    return {
      drive: {
        'style/': { 'action_buttons.css': { '$ref': 'action_buttons.css' } },
        'data/': { 'opts.json': { raw: opts  || {}} }
      },
      _: { general_button, send_btc, receive_btc }
    }
  }
}
