// lightning_buttons/lightning_buttons.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { sdb, get } = statedb(fallback_module)

const general_button = require('general_button')
const pay_invoice = require('pay_invoice')
const create_invoice = require('create_invoice')

module.exports = lightning_buttons

async function lightning_buttons(opts = {}, protocol) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = { style: inject, data: ondata }
  const _ = { pay_general: null, create_general: null }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  shadow.innerHTML = `
    <div class="lightning-buttons-container">
      <div class="pay-create-buttons">
        <div id="pay-button-container"></div>
        <div id="create-button-container"></div>
      </div>
    </div>
    <div class="overlay pay-overlay hidden"></div>
    <div class="overlay create-overlay hidden"></div>
    <style></style>
  `

  const style = shadow.querySelector('style')
  const pay_container = shadow.querySelector('#pay-button-container')
  const create_container = shadow.querySelector('#create-button-container')
  const pay_overlay = shadow.querySelector('.pay-overlay')
  const create_overlay = shadow.querySelector('.create-overlay')

  const subs = await sdb.watch(onbatch)

  // -------------------- Protocol --------------------
  let sendUp = () => {}
  if (protocol) {
    sendUp = protocol(msg => console.log('[lightning_buttons]:', msg))
  }

  // -------------------- Buttons --------------------
  const pay_button = await general_button(subs[0], button_protocol('pay_general'))
  const create_button = await general_button(subs[1], button_protocol('create_general'))

  pay_container.appendChild(pay_button)
  create_container.appendChild(create_button)

  pay_button._action = 'pay_message'
  create_button._action = 'create_message'

  let pay_el = null, create_el = null

  function close_all_overlays(except) {
    if (except !== 'pay') pay_overlay.classList.add('hidden')
    if (except !== 'create') create_overlay.classList.add('hidden')
  }

  // -------------------- Button Clicks --------------------
  pay_button.onclick = async e => {
    e.stopPropagation()
    close_all_overlays('pay')
    if (!pay_el) {
      pay_el = await pay_invoice(subs[2], {
        onClose: () => pay_overlay.classList.add('hidden')
      })
      pay_overlay.appendChild(pay_el)
    }
    pay_overlay.classList.remove('hidden')
    sendUp({ from: 'lightning_buttons', type: 'pay', data: null })
  }

  create_button.onclick = async e => {
    e.stopPropagation()
    close_all_overlays('create')
    if (!create_el) {
      create_el = await create_invoice(subs[3], {
        onClose: () => create_overlay.classList.add('hidden')
      })
      create_overlay.appendChild(create_el)
    }
    create_overlay.classList.remove('hidden')
    sendUp({ from: 'lightning_buttons', type: 'create', data: null })
  }

  // -------------------- Initial Config --------------------
  _.pay_general?.({ type: 'button_name', data: { name: 'Pay Invoice', action: 'pay_message' } })
  _.create_general?.({ type: 'button_name', data: { name: 'Create Invoice', action: 'create_message' } })

  return el

  // -------------------- Helpers --------------------
  function fail(data, type) { throw new Error('Invalid message type', { cause: { data, type } }) }
  async function onbatch(batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(p => drive.get(p).then(f => f.raw)))
      const handler = on[type] || fail
      handler(data, type)
    }
  }
  function inject(data) { style.textContent = data[0] }
  async function ondata(data) { const buttonData = data[0]?.value || {} }
  function button_protocol(key) { return send => { _[key] = send; return send } }
}

// ------------------ Fallback ------------------
function fallback_module() {
  return {
    api,
    _: { 'general_button': { $: '' }, 'pay_invoice': { $: '' }, 'create_invoice': { $: '' } }
  }

  function api(opts) {
    if (!opts) opts = {}
    if (!opts.value) opts.value = {}
 
    const general_button = { mapping: { style: 'style', data: 'data' }, 0: {}, 1: {} }
    const pay_invoice = { mapping: { style: 'style', data: 'data', icons: 'icons' }, 2: '' }
    const create_invoice = { mapping: { style: 'style', data: 'data', icons: 'icons' }, 3: '' }

    return {
      drive: {
        'style/': { 'lightning_buttons.css': { '$ref': 'lightning_buttons.css' } },
        'data/': { 'opts.json': { raw: opts || {} } }
      },
      _: { general_button, pay_invoice, create_invoice }
    }
  }
}
