// receipt_row/receipt_row.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { sdb, get } = statedb(fallback_module)

module.exports = receipt_row

const btc_usd_rate = require('btc_usd_rate')

async function receipt_row(opts = {}) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = {
    style: inject,
    data: ondata,
    icons: iconject,
  }

  let dricons = []
  let current_unit = 'BTC'
  let btc_to_usd = await btc_usd_rate()

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  shadow.innerHTML = `
    <div class="receipt-row"></div>
    <style></style>
  `

  const style = shadow.querySelector('style')
  const row = shadow.querySelector('.receipt-row')

  await sdb.watch(onbatch)

  return el

  function fail(data, type) {
    throw new Error('invalid message', { cause: { data, type } })
  }

  async function onbatch(batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(path => drive.get(path).then(file => file.raw)))
      const func = on[type] || fail
      await func(data, type)
    }
  }

  function inject(data) {
    style.textContent = data[0]
  }

  function iconject(data) {
    dricons = data
  }

  async function ondata(data) {
    let { label, value, link, icon, convert } = data[0] || {}

    let value_html = value || ""

    if (link) {
      value_html = `<a href="${value}" target="_blank" class="receipt-link">${value}</a>`
      row.className = `receipt-row link`
    }

    if (icon) {
      if (icon == "btc.svg") {
        value_html = `<span class="receipt-icon">${dricons[1]}</span> ${value}`
        row.className = `receipt-row total`
      } else if (icon == "lightning.svg") {
        value_html = `<span class="receipt-icon">${dricons[2]}</span> ${value}`
        row.className = `receipt-row total`
      } else {
        value_html = `<span class="receipt-icon">${dricons[0]}</span> ${value}`
      }
    }

    // Base layout
    row.innerHTML = `
      <div class="receipt-label">${label}</div>
      <div class="receipt-bottom">
        <div class="receipt-value">${value_html}</div>
        ${convert ? `
          <div class="convert-buttons">
            <button class="convert-btn btc active">BTC</button>
            <button class="convert-btn usd">USD</button>
          </div>` : ""}
      </div>
      ${icon ? "" : `<div class="divider"></div>`}
    `

    if (convert) {
      const btc_btn = row.querySelector('.convert-btn.btc')
      const usd_btn = row.querySelector('.convert-btn.usd')
      const value_el = row.querySelector('.receipt-value')

      btc_btn.onclick = () => {
        current_unit = 'BTC'
        btc_btn.classList.add('active')
        usd_btn.classList.remove('active')
        value_el.innerHTML = `${value}`
      }

      usd_btn.onclick = () => {
        current_unit = 'USD'
        usd_btn.classList.add('active')
        btc_btn.classList.remove('active')

        let btc_value = parseFloat(value)
        if (isNaN(btc_value)) btc_value = 0

        let usd_value = (btc_value * btc_to_usd).toFixed(2)
        value_el.innerHTML = `$${usd_value}`
      }
    }
  }
}

function fallback_module() {
  return {
    api: fallback_instance
  }
  function fallback_instance(opts) {
    return {
      drive: {
        'icons/': {
          'default.svg': { '$ref': 'default.svg' },
          'btc.svg': { '$ref': 'btc.svg' },
          'lightning.svg': { '$ref': 'lightning.svg' },
        },
        'style/': {
          'receipt_row.css': {
           '$ref':'receipt_row.css'
          }
        },
        'data/': {
          'opts.json': { raw: opts }
        }
      }
    }
  }
}
